// @tma.js/sdk/dist/index.js
var nt = Object.defineProperty;
var it = (s, t, r) => t in s ? nt(s, t, { enumerable: !0, configurable: !0, writable: !0, value: r }) : s[t] = r;
var e = (s, t, r) => it(s, typeof t != "symbol" ? t + "" : t, r);
import { batch as ce, signal as j, computed as C } from "@tma.js/signals";
import { supports as ee, isTMAFp as ot, postEventFp as at, off as ne, on as z, requestFp as pt, request2Fp as ut, invokeCustomMethodFp as ct, retrieveLaunchParamsFp as le, retrieveRawInitDataFp as lt, captureSameReq as ht, logger as Re } from "@tma.js/bridge";
import { CancelledError as vn, InvalidLaunchParamsError as kn, InvokeCustomMethodFailedError as In, LaunchParamsRetrieveError as yn, MethodParameterUnsupportedError as En, MethodUnsupportedError as qn, TimeoutError as An, UnknownEnvError as Bn, applyPolyfills as xn, createLogger as Mn, createPostEvent as Vn, createStartParam as Tn, createStartParamFp as Ln, debug as $n, decodeBase64Url as Rn, decodeBase64UrlFp as Pn, decodeStartParam as jn, decodeStartParamFp as On, deepSnakeToCamelObjKeys as Dn, emitEvent as Gn, encodeBase64Url as Un, getReleaseVersion as Wn, isSafeToCreateStartParam as Hn, isTMA as Kn, isTMAFp as Qn, logger as zn, mockTelegramEnv as Nn, off as Jn, offAll as Yn, on as Zn, retrieveLaunchParams as Xn, retrieveLaunchParamsFp as ei, retrieveRawInitData as ti, retrieveRawInitDataFp as si, retrieveRawLaunchParams as ri, retrieveRawLaunchParamsFp as ni, setDebug as ii, setTargetOrigin as oi, supports as ai, targetOrigin as pi } from "@tma.js/bridge";
import { throwifyFpFn as te, throwifyAnyEither as De, getStorageValue as dt, setStorageValue as _t, snakeToKebab as mt, createCbCollector as ke, camelToKebab as Ge, BetterTaskEither as ge } from "@tma.js/toolkit";
import { toRGBFullFp as ft, isRGB as _e, pipeQueryToSchema as bt, pipeJsonToSchema as gt } from "@tma.js/transformers";
import { isAnyRGB as ci, isRGB as li, isRGBA as hi, isRGBAShort as di, isRGBShort as _i, parseInitDataQuery as mi, parseInitDataQueryFp as fi, parseLaunchParamsQuery as bi, parseLaunchParamsQueryFp as gi, serializeInitDataQuery as Fi, serializeLaunchParamsQuery as Ci, toRGB as wi, toRGBFp as Si, toRGBFull as vi, toRGBFullFp as ki } from "@tma.js/transformers";
import { either as b, function as u, option as Q, taskEither as l } from "fp-ts";
import { errorClassWithData as Ie, errorClass as D } from "error-kid";
import { parse as Pe, record as Ft, string as K, array as Ct, safeParse as Fe, looseObject as je, pipe as Ue, transform as We, date as He, optional as wt, number as Ke, integer as St } from "valibot";
function V(s) {
  return typeof s == "function" ? s() : s;
}
class ie {
  constructor({
    onMounted: t,
    restoreState: r,
    initialState: n,
    onUnmounted: a,
    isPageReload: p
  }) {
    e(this, "_isMounted", j(!1));
    /**
     * Signal indicating if the component is mounted.
     */
    e(this, "isMounted", C(this._isMounted));
    /**
     * Mounts the component restoring its state and calling required side effects.
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     */
    e(this, "unmount");
    this.mount = () => {
      if (this.isMounted())
        return b.right(void 0);
      const o = V(p) ? r() : void 0, i = o ? b.right(o) : typeof n == "function" ? n() : b.right(n);
      return u.pipe(i, b.map((c) => {
        ce(() => {
          this._isMounted.set(!0), t == null || t(c);
        });
      }));
    }, this.unmount = () => {
      this._isMounted() && ce(() => {
        this._isMounted.set(!1), a == null || a();
      });
    };
  }
}
function Ce(s) {
  const t = {};
  for (const r in s) {
    const n = s[r];
    n !== void 0 && (t[r] = n);
  }
  return t;
}
function me(s, t) {
  const r = Object.keys(s), n = Object.keys(t);
  return r.length !== n.length ? !1 : r.every((a) => Object.prototype.hasOwnProperty.call(t, a) && s[a] === t[a]);
}
class J {
  constructor({ initialState: t, onChange: r }) {
    e(this, "_state");
    /**
     * The current state.
     */
    e(this, "state");
    /**
     * Updates the state.
     * @param state - updates to apply.
     */
    e(this, "setState", (t) => {
      const r = { ...this.state(), ...Ce(t) };
      me(r, this.state()) || this._state.set(r);
    });
    this._state = j(t, { equals: me }), this.state = C(this._state), this.state.sub(r);
  }
  /**
   * Creates a computed signal based on the state.
   * @param key - a state key to use as a source.
   */
  getter(t) {
    return C(() => this._state()[t]);
  }
  /**
   * @returns True if specified payload will update the state.
   * @param state
   */
  hasDiff(t) {
    return !me({ ...this.state(), ...Ce(t) }, this.state());
  }
}
function W(s, t) {
  return C(() => ee(s, V(t)));
}
// @__NO_SIDE_EFFECTS__
function h(s) {
  return Object.assign(te(s), {
    ifAvailable(...t) {
      return u.pipe(
        s.ifAvailable(...t),
        Q.match(
          () => ({ ok: !1 }),
          (r) => ({
            ok: !0,
            data: De(r)
          })
        )
      );
    }
  });
}
function X(s) {
  return [s];
}
class ue extends (/* @__PURE__ */ Ie({
  name: "ValidationError",
  data: (t, r) => ({ input: t, issues: r }),
  super: "Validation error"
})) {
}
class ye extends (/* @__PURE__ */ D({
  name: "CSSVarsBoundError",
  super: "CSS variables are already bound"
})) {
}
class vt extends (/* @__PURE__ */ Ie({
  name: "DeviceStorageMethodError",
  data: (t) => ({ error: t }),
  super: (t) => [t]
})) {
}
class kt extends (/* @__PURE__ */ Ie({
  name: "SecureStorageMethodError",
  data: (t) => ({ error: t }),
  super: (t) => [t]
})) {
}
class It extends (/* @__PURE__ */ D({
  name: "NotAvailableError",
  super: X
})) {
}
class Mr extends (/* @__PURE__ */ D({
  name: "InvalidEnvError",
  super: X
})) {
}
class yt extends (/* @__PURE__ */ D({
  name: "FunctionNotAvailableError",
  super: X
})) {
}
class U extends (/* @__PURE__ */ D({
  name: "InvalidArgumentsError",
  super: (t, r) => [t, { cause: r }]
})) {
}
class Ee extends (/* @__PURE__ */ D({
  name: "ConcurrentCallError",
  super: X
})) {
}
class Et extends (/* @__PURE__ */ D({
  name: "SetEmojiStatusError",
  super: (t) => [`Failed to set emoji status: ${t}`]
})) {
}
class Qe extends (/* @__PURE__ */ D({
  name: "AccessDeniedError",
  super: X
})) {
}
class qt extends (/* @__PURE__ */ D({
  name: "FullscreenFailedError",
  super: X
})) {
}
class At extends (/* @__PURE__ */ D({
  name: "ShareMessageError",
  super: X
})) {
}
class Bt extends (/* @__PURE__ */ D({
  name: "UnknownThemeParamsKeyError",
  super: (t) => [`Unknown theme params key passed: ${t}`]
})) {
}
function A(s, t) {
  const r = C(() => V(t.version) || "100"), n = C(() => V(t.isTma)), { requires: a, returns: p } = t, o = a ? typeof a == "object" ? a : { every: [a] } : void 0, i = (f) => {
    if (!t.supports)
      return !0;
    const g = t.supports[f];
    return ee(g.method, g.param, r());
  }, c = () => {
    if (!o)
      return;
    const [f, g] = "every" in o ? ["every", o.every] : ["some", o.some];
    for (let d = 0; d < g.length; d++) {
      const B = g[d], L = typeof B == "function" ? B() : ee(B, r()) ? void 0 : `it is unsupported in Mini Apps version ${r()}`;
      if (L && (f === "every" || d === g.length - 1))
        return L;
    }
  }, m = (...f) => {
    for (const g in t.supports)
      if (t.supports[g].shouldCheck(...f) && !i(g))
        return `option ${g} is not supported in Mini Apps version ${r()}`;
  }, F = C(() => !c()), _ = C(() => r() !== "0.0"), S = C(() => t.isMounted ? t.isMounted() : !0), w = C(
    () => n() && _() && F() && S()
  ), k = (f) => {
    const g = new yt(f);
    return ["task", "promise"].includes(t.returns) ? l.left(g) : b.left(g);
  }, I = (...f) => p === "plain" ? b.tryCatch(() => s(...f), (g) => g) : p === "promise" ? l.tryCatch(() => s(...f), (g) => g) : s(...f);
  return Object.assign(
    (...f) => {
      var L;
      const g = "Unable to call function:";
      if (!n())
        return k(`${g} it can't be called outside Mini Apps`);
      if (!_())
        return k(`${g} the SDK was not initialized. Use the SDK init() function`);
      const d = c();
      if (d)
        return k(`${g} ${d}`);
      const B = m(...f);
      if (B)
        return k(`${g} ${B}`);
      if (!S()) {
        const M = (L = t.isMounting) != null && L.call(t) ? "mounting. Wait for the mount completion" : "unmounted. Use the mount() method";
        return k(`${g} the component is ${M}`);
      }
      return I(...f);
    },
    s,
    {
      isAvailable: w,
      ifAvailable(...f) {
        return w() ? Q.some(I(...f)) : Q.none;
      }
    },
    o ? { isSupported: F } : {},
    t.supports ? { supports: i } : {}
  );
}
function E(s) {
  return (t) => A(t, s);
}
class de {
  constructor({
    isTma: t,
    storage: r,
    onClick: n,
    offClick: a,
    initialState: p,
    isPageReload: o,
    postEvent: i,
    payload: c,
    method: m,
    version: F
  }) {
    /**
     * Signal indicating if the component is currently mounted.
     */
    e(this, "isMounted");
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * Complete button state.
     */
    e(this, "state");
    /**
     * @returns A setter with checks for the specified key.
     * @param key
     */
    e(this, "stateSetters");
    /**
     * @returns Setters with checks to set a specified boolean key.
     * @param key
     */
    e(this, "stateBoolSetters");
    /**
     * Updates the button state.
     */
    e(this, "setStateFp");
    /**
     * @see setStateFp
     */
    e(this, "setState");
    /**
     * Adds a new button listener.
     * @param listener - event listener.
     * @param once - should the listener be called only once.
     * @returns A function to remove bound listener.
     * @example
     * const off = button.onClick(() => {
     *   console.log('User clicked the button');
     *   off();
     * });
     */
    e(this, "onClickFp");
    /**
     * @see onClickFp
     */
    e(this, "onClick");
    /**
     * Removes the button click listener.
     * @param listener - event listener.
     * @param once - should the listener be called only once.
     * @example
     * function listener() {
     *   console.log('User clicked the button');
     *   button.offClick(listener);
     * }
     * button.onClick(listener);
     */
    e(this, "offClickFp");
    /**
     * @see offClickFp
     */
    e(this, "offClick");
    /**
     * Mounts the component restoring its state.
     * @since Mini Apps v6.1
     */
    e(this, "mountFp");
    /**
     * @see mountFp
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     *
     * Note that this function does not remove listeners added via the `onClick`
     * function, so you have to remove them on your own.
     * @see onClick
     */
    e(this, "unmount");
    const _ = new J({
      initialState: p,
      onChange(f) {
        r.set(f);
      }
    }), S = new ie({
      initialState: p,
      isPageReload: o,
      onMounted: _.setState,
      restoreState: r.get
    }), w = { version: F, requires: m, isTma: t }, k = E({
      ...w,
      returns: "plain"
    }), I = E({
      ...w,
      returns: "either",
      isMounted: S.isMounted
    });
    this.isMounted = S.isMounted, this.isSupported = W(m, F), this.state = _.state, this.setStateFp = I((f) => {
      const g = { ...this.state(), ...Ce(f) };
      return _.hasDiff(g) ? u.pipe(
        i(m, c(g)),
        b.map(() => {
          _.setState(g);
        })
      ) : b.right(void 0);
    }), this.setState = /* @__PURE__ */ h(this.setStateFp), this.onClickFp = k(n), this.onClick = /* @__PURE__ */ h(this.onClickFp), this.offClickFp = k(a), this.offClick = /* @__PURE__ */ h(this.offClickFp), this.mountFp = k(() => {
      const f = () => {
      };
      return u.pipe(S.mount(), b.match(f, f));
    }), this.mount = /* @__PURE__ */ h(this.mountFp), this.unmount = S.unmount, this.stateSetters = (f) => {
      const g = I((d) => this.setStateFp({ [f]: d }));
      return [/* @__PURE__ */ h(g), g];
    }, this.stateBoolSetters = (f) => {
      const [, g] = this.stateSetters(f), d = I(() => g(!1)), B = I(() => g(!0));
      return [
        [/* @__PURE__ */ h(d), d],
        [/* @__PURE__ */ h(B), B]
      ];
    };
  }
  /**
   * @returns A computed based on the specified state and its related key.
   * @param key - a key to use.
   */
  stateGetter(t) {
    return C(() => this.state()[t]);
  }
}
class xt {
  constructor(t) {
    /**
     * Signal indicating if the component is currently visible.
     */
    e(this, "isVisible");
    /**
     * Signal indicating if the component is currently mounted.
     */
    e(this, "isMounted");
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * Hides the button.
     * @since Mini Apps v6.1
     */
    e(this, "hideFp");
    /**
     * @see hideFp
     */
    e(this, "hide");
    /**
     * Shows the button.
     * @since Mini Apps v6.1
     */
    e(this, "showFp");
    /**
     * @see showFp
     */
    e(this, "show");
    /**
     * Adds a new button listener.
     * @param listener - event listener.
     * @param once - should the listener be called only once.
     * @returns A function to remove bound listener.
     * @since Mini Apps v6.1
     * @example
     * const off = button.onClick(() => {
     *   console.log('User clicked the button');
     *   off();
     * });
     */
    e(this, "onClickFp");
    /**
     * @see onClickFp
     */
    e(this, "onClick");
    /**
     * Removes the button click listener.
     * @param listener - event listener.
     * @param once - should the listener be called only once.
     * @since Mini Apps v6.1
     * @example
     * function listener() {
     *   console.log('User clicked the button');
     *   button.offClick(listener);
     * }
     * button.onClick(listener);
     */
    e(this, "offClickFp");
    /**
     * @see offClickFp
     */
    e(this, "offClick");
    /**
     * Mounts the component restoring its state.
     * @since Mini Apps v6.1
     */
    e(this, "mountFp");
    /**
     * @see mountFp
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     *
     * Note that this function does not remove listeners added via the `onClick`
     * function, so you have to remove them on your own.
     * @see onClick
     */
    e(this, "unmount");
    const r = new de({
      ...t,
      method: "web_app_setup_back_button",
      payload: (n) => ({ is_visible: n.isVisible }),
      initialState: { isVisible: !1 }
    });
    this.isVisible = r.stateGetter("isVisible"), this.isMounted = r.isMounted, this.isSupported = r.isSupported, [[this.hide, this.hideFp], [this.show, this.showFp]] = r.stateBoolSetters("isVisible"), this.onClick = r.onClick, this.onClickFp = r.onClickFp, this.offClick = r.offClick, this.offClickFp = r.offClickFp, this.mount = r.mount, this.mountFp = r.mountFp, this.unmount = r.unmount;
  }
}
function v() {
  return { isTma: C(() => ot()) };
}
function se(s) {
  return (t) => ({ ...t, ...V(s) });
}
// @__NO_SIDE_EFFECTS__
function oe(s, t) {
  return j(s, t);
}
const ze = /* @__PURE__ */ oe(at), re = (...s) => ze()(...s), Mt = (...s) => De(re(...s)), T = se({
  postEvent: re
});
function Vt(s) {
  return {
    get: () => dt(s),
    set(t) {
      _t(s, t);
    }
  };
}
function Tt() {
  return performance.getEntriesByType("navigation")[0];
}
function Lt() {
  const s = Tt();
  return !!s && s.type === "reload";
}
function Y(s) {
  return se({
    storage: Vt(s),
    isPageReload: Lt
  });
}
const we = /* @__PURE__ */ oe("0.0"), y = se({ version: we });
// @__NO_SIDE_EFFECTS__
function qe(s, t) {
  return {
    ...u.pipe(
      v(),
      T,
      y,
      Y(s)
    ),
    onClick(r, n) {
      return z(t, r, n);
    },
    offClick(r, n) {
      ne(t, r, n);
    }
  };
}
const Vr = /* @__PURE__ */ new xt(
  /* @__PURE__ */ qe("backButton", "back_button_pressed")
);
class Ae {
  constructor({
    initialState: t,
    onMounted: r,
    restoreState: n,
    onUnmounted: a,
    isPageReload: p
  }) {
    e(this, "_isMounted", j(!1));
    /**
     * Signal indicating if the component is mounted.
     */
    e(this, "isMounted", C(this._isMounted));
    /**
     * Mounts the component restoring its state and calling required side effects.
     * @param options - additional execution options.
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     */
    e(this, "unmount");
    this.mount = (o) => {
      if (this._isMounted())
        return l.right(void 0);
      const i = V(p) ? n() : void 0;
      return u.pipe(
        i ? l.right(i) : t(o),
        l.map((c) => {
          this._isMounted() || ce(() => {
            this._isMounted.set(!0), r == null || r(c);
          });
        })
      );
    }, this.unmount = () => {
      this._isMounted() && ce(() => {
        this._isMounted.set(!1), a == null || a();
      });
    };
  }
}
function Oe() {
  return new It("Biometry is not available");
}
function fe(s) {
  let t = !1, r = !1, n = "", a = !1, p = "", o = !1;
  return s.available && (t = !0, r = s.token_saved, n = s.device_id, a = s.access_requested, p = s.type, o = s.access_granted), { available: t, tokenSaved: r, deviceId: n, type: p, accessGranted: o, accessRequested: a };
}
class $t {
  constructor({
    version: t,
    request: r,
    postEvent: n,
    storage: a,
    onInfoReceived: p,
    offInfoReceived: o,
    isTma: i,
    isPageReload: c
  }) {
    /**
     * Signal indicating if biometry is available.
     */
    e(this, "isAvailable");
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * Signal indicating if the component is mounted.
     */
    e(this, "isMounted");
    /**
     * Complete component state.
     */
    e(this, "state");
    /**
     * Attempts to authenticate a user using biometrics and fetch a previously stored secure token.
     * @param options - method options.
     * @since Mini Apps v7.2
     * @returns Token from the local secure storage saved previously or undefined.
     * @example
     * const { status, token } = await biometry.authenticate({
     *   reason: 'Authenticate to open wallet',
     * });
     */
    e(this, "authenticateFp");
    /**
     * @see authenticateFp
     */
    e(this, "authenticate");
    /**
     * Opens the biometric access settings for bots. Useful when you need to request biometrics
     * access to users who haven't granted it yet.
     *
     * _Note that this method can be called only in response to user interaction with the Mini App
     * interface (e.g. a click inside the Mini App or on the main button)_.
     * @since Mini Apps v7.2
     */
    e(this, "openSettingsFp");
    /**
     * @see openSettingsFp
     */
    e(this, "openSettings");
    /**
     * Requests permission to use biometrics.
     * @since Mini Apps v7.2
     * @returns Promise with true, if access was granted.
     * @example
     * const accessGranted = await biometry.requestAccess({
     *   reason: 'Authenticate to open wallet',
     * });
     */
    e(this, "requestAccessFp");
    /**
     * @see requestAccessFp
     */
    e(this, "requestAccess");
    /**
     * Updates the biometric token in a secure storage on the device.
     * @since Mini Apps v7.2
     * @returns Promise with `true`, if token was updated.
     * @example Setting a new token
     * biometry.updateToken({
     *   token: 'abcdef',
     * })
     * @example Deleting the token
     * biometry.updateToken();
     */
    e(this, "updateTokenFp");
    /**
     * @see updateTokenFp
     */
    e(this, "updateToken");
    /**
     * Mounts the component restoring its state.
     * @since Mini Apps v7.2
     */
    e(this, "mountFp");
    /**
     * @see mountFp
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     */
    e(this, "unmount");
    const m = (f) => {
      F.setState(fe(f));
    }, F = new J({
      initialState: {
        available: !1,
        type: "unknown",
        accessGranted: !1,
        accessRequested: !1,
        deviceId: "",
        tokenSaved: !1
      },
      onChange: a.set
    }), _ = new Ae({
      initialState(f) {
        return u.pipe(
          r("web_app_biometry_get_info", "biometry_info_received", f),
          l.map(fe)
        );
      },
      isPageReload: c,
      onMounted(f) {
        F.setState(f), p(m);
      },
      onUnmounted() {
        o(m);
      },
      restoreState: a.get
    }), S = { version: t, requires: "web_app_biometry_request_auth", isTma: i }, w = E({
      ...S,
      returns: "either"
    }), k = E({
      ...S,
      returns: "task"
    }), I = E({
      ...S,
      isMounted: _.isMounted,
      returns: "task"
    });
    this.isAvailable = F.getter("available"), this.isMounted = _.isMounted, this.isSupported = W("web_app_biometry_request_auth", t), this.state = F.state, this.unmount = _.unmount, this.mountFp = k(_.mount), this.authenticateFp = I((f) => this.isAvailable() ? u.pipe(
      r("web_app_biometry_request_auth", "biometry_auth_requested", {
        ...f,
        params: { reason: ((f || {}).reason || "").trim() }
      }),
      l.map((g) => (F.setState({ token: g.token }), g))
    ) : l.left(Oe())), this.openSettingsFp = w(() => n("web_app_biometry_open_settings")), this.requestAccessFp = I((f) => u.pipe(
      r("web_app_biometry_request_access", "biometry_info_received", {
        ...f,
        params: { reason: ((f || {}).reason || "").trim() }
      }),
      l.chain((g) => {
        const d = fe(g);
        return d.available ? (F.setState(d), l.right(d.accessRequested)) : l.left(Oe());
      })
    )), this.updateTokenFp = I((f = {}) => {
      var g;
      return u.pipe(
        r("web_app_biometry_update_token", "biometry_token_updated", {
          ...f,
          params: { token: f.token || "", reason: (g = f.reason) == null ? void 0 : g.trim() }
        }),
        l.map((d) => d.status)
      );
    }), this.authenticate = /* @__PURE__ */ h(this.authenticateFp), this.openSettings = /* @__PURE__ */ h(this.openSettingsFp), this.requestAccess = /* @__PURE__ */ h(this.requestAccessFp), this.updateToken = /* @__PURE__ */ h(this.updateTokenFp), this.mount = /* @__PURE__ */ h(this.mountFp);
  }
}
const Rt = (s, t, r) => pt(s, t, {
  postEvent: re,
  ...r
}), Ne = (s, t, r) => ut(s, t, {
  postEvent: re,
  ...r
}), Tr = te(Rt), Lr = te(Ne), $ = se({ request: Ne });
function Pt() {
  return new $t({
    ...u.pipe(
      v(),
      T,
      y,
      $,
      Y("biometry")
    ),
    offInfoReceived(s) {
      ne("biometry_info_received", s);
    },
    onInfoReceived(s) {
      return z("biometry_info_received", s);
    }
  });
}
const $r = /* @__PURE__ */ Pt();
class jt {
  constructor({ postEvent: t, storage: r, isTma: n, isPageReload: a }) {
    /**
     * Signal indicating if closing confirmation dialog is currently enabled.
     */
    e(this, "isConfirmationEnabled");
    /**
     * Signal indicating if the component is currently mounted.
     */
    e(this, "isMounted");
    /**
     * Mounts the component restoring its state.
     */
    e(this, "mountFp");
    /**
     * @see mountFp
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     */
    e(this, "unmount");
    /**
     * Disables the closing confirmation dialog.
     */
    e(this, "disableConfirmationFp");
    /**
     * @see disableConfirmationFp
     */
    e(this, "disableConfirmation");
    /**
     * Enables the closing confirmation dialog.
     */
    e(this, "enableConfirmationFp");
    /**
     * @see enableConfirmationFp
     */
    e(this, "enableConfirmation");
    const p = new J({
      initialState: { isConfirmationEnabled: !1 },
      onChange(_) {
        r.set(_);
      }
    }), o = new ie({
      onMounted: p.setState,
      restoreState: r.get,
      initialState: { isConfirmationEnabled: !1 },
      isPageReload: a
    }), i = { requires: "web_app_setup_closing_behavior", isTma: n }, c = E({
      ...i,
      returns: "plain"
    }), m = E({
      ...i,
      returns: "either",
      isMounted: o.isMounted
    }), F = (_) => _ === this.isConfirmationEnabled() ? b.right(void 0) : (p.setState({ isConfirmationEnabled: _ }), t("web_app_setup_closing_behavior", {
      need_confirmation: _
    }));
    this.isConfirmationEnabled = p.getter("isConfirmationEnabled"), this.isMounted = o.isMounted, this.disableConfirmationFp = m(() => F(!1)), this.enableConfirmationFp = m(() => F(!0)), this.mountFp = c(() => {
      const _ = () => {
      };
      return u.pipe(o.mount(), b.match(_, _));
    }), this.unmount = o.unmount, this.disableConfirmation = /* @__PURE__ */ h(this.disableConfirmationFp), this.enableConfirmation = /* @__PURE__ */ h(this.enableConfirmationFp), this.mount = /* @__PURE__ */ h(this.mountFp);
  }
}
function Ot() {
  return new jt(u.pipe(
    v(),
    Y("closingBehavior"),
    T
  ));
}
const Rr = /* @__PURE__ */ Ot();
class Dt {
  constructor({ version: t, isTma: r, invokeCustomMethod: n }) {
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * Deletes specified key or keys from the cloud storage.
     * @param keyOrKeys - key or keys to delete.
     * @param options - request execution options.
     * @since Mini Apps v6.9
     */
    e(this, "deleteItemFp");
    e(this, "deleteItem");
    /**
     * Gets a single key value from the cloud storage.
     * @param key - a key to get.
     * @param options - request execution options.
     * @returns A key value as a string.
     * @since Mini Apps v6.9
     */
    e(this, "getItemFp");
    /**
     * @see getItemFp
     */
    e(this, "getItem");
    /**
     * Gets multiple keys' values from the cloud storage.
     * @param keys - keys list.
     * @param options - request execution options.
     * @returns A map, where a key is one of the specified in the `keys` argument,
     * and a value is a corresponding storage value if an array of keys was passed.
     * @since Mini Apps v6.9
     */
    e(this, "getItemsFp");
    /**
     * @see getItemsFp
     */
    e(this, "getItems");
    /**
     * Returns a list of all keys presented in the cloud storage.
     * @param options - request execution options.
     * @since Mini Apps v6.9
     */
    e(this, "getKeysFp");
    /**
     * @see getKeysFp
     */
    e(this, "getKeys");
    /**
     * Saves the specified value by a key.
     * @param key - storage key.
     * @param value - storage value.
     * @param options - request execution options.
     * @since Mini Apps v6.9
     */
    e(this, "setItemFp");
    /**
     * @see setItemFp
     */
    e(this, "setItem");
    /**
     * Clears the cloud storage.
     * @param options - additional options.
     * @since Mini Apps v6.9
     */
    e(this, "clearFp");
    /**
     * @see clearFp
     */
    e(this, "clear");
    const a = E({
      version: t,
      requires: "web_app_invoke_custom_method",
      isTma: r,
      returns: "task"
    });
    this.isSupported = W("web_app_invoke_custom_method", t), this.deleteItemFp = a((p, o) => {
      const i = Array.isArray(p) ? p : [p];
      return u.pipe(
        i.length ? n("deleteStorageValues", { keys: i }, o) : l.right(void 0),
        l.map(() => {
        })
      );
    }), this.getItemFp = a((p, o) => u.pipe(
      this.getItemsFp([p], o),
      l.map((i) => i[p] || "")
    )), this.getItemsFp = a((p, o) => u.pipe(
      p.length ? n("getStorageValues", { keys: p }, o) : l.right({}),
      l.map((i) => ({
        // Fulfill the response with probably missing keys.
        ...p.reduce((c, m) => (c[m] = "", c), {}),
        ...Pe(Ft(K(), K()), i)
      }))
    )), this.getKeysFp = a((p) => u.pipe(
      n("getStorageKeys", {}, p),
      l.map((o) => Pe(Ct(K()), o))
    )), this.setItemFp = a((p, o, i) => u.pipe(
      n("saveStorageValue", { key: p, value: o }, i),
      l.map(() => {
      })
    )), this.clearFp = a((p) => u.pipe(this.getKeysFp(p), l.chain(this.deleteItemFp))), this.deleteItem = /* @__PURE__ */ h(this.deleteItemFp), this.getItem = /* @__PURE__ */ h(this.getItemFp), this.getItems = /* @__PURE__ */ h(this.getItemsFp), this.getKeys = /* @__PURE__ */ h(this.getKeysFp), this.setItem = /* @__PURE__ */ h(this.setItemFp), this.clear = /* @__PURE__ */ h(this.clearFp);
  }
}
const be = /* @__PURE__ */ oe(0);
function Be() {
  return be.set(be() + 1), be().toString();
}
function Gt(s, t, r) {
  return ct(s, t, Be(), {
    ...r || {},
    postEvent: re
  });
}
const xe = se({
  invokeCustomMethod: Gt
});
function Ut() {
  return new Dt(u.pipe(
    v(),
    y,
    xe
  ));
}
const Pr = /* @__PURE__ */ Ut();
class Wt {
  constructor({ isTma: t, request: r, version: n, createRequestId: a }) {
    /**
      * Retrieves an item using its key.
      * @since Mini Apps v9.0
      */
    e(this, "getItemFp");
    /**
     * @see getItemFp
     */
    e(this, "getItem");
    /**
      * Sets a new item in the storage.
      * @since Mini Apps v9.0
      */
    e(this, "setItemFp");
    /**
     * @see setItemFp
     */
    e(this, "setItem");
    /**
      * Removes a key from the storage.
      * @since Mini Apps v9.0
      */
    e(this, "deleteItemFp");
    /**
     * @see deleteItemFp
     */
    e(this, "deleteItem");
    /**
      * Removes all keys from the storage.
      * @since Mini Apps v9.0
      */
    e(this, "clearFp");
    /**
     * @see clearFp
     */
    e(this, "clear");
    const p = E({
      version: n,
      requires: "web_app_device_storage_get_key",
      isTma: t,
      returns: "task"
    }), o = (i, c, m) => {
      const F = a();
      return u.pipe(
        r(i, ["device_storage_failed", c], {
          params: { ...m, req_id: F },
          capture: (_) => "payload" in _ ? _.payload.req_id === F : !0
        }),
        l.chain((_) => _.event === "device_storage_failed" ? l.left(new vt(_.payload.error || "UNKNOWN_ERROR")) : l.right(_.payload))
      );
    };
    this.getItemFp = p((i) => u.pipe(
      o("web_app_device_storage_get_key", "device_storage_key_received", { key: i }),
      l.map((c) => c.value)
    )), this.setItemFp = p((i, c) => u.pipe(
      o("web_app_device_storage_save_key", "device_storage_key_saved", { key: i, value: c }),
      l.map(() => {
      })
    )), this.deleteItemFp = p((i) => this.setItemFp(i, null)), this.clearFp = p(() => u.pipe(
      o("web_app_device_storage_clear", "device_storage_cleared", {}),
      l.map(() => {
      })
    )), this.getItem = /* @__PURE__ */ h(this.getItemFp), this.setItem = /* @__PURE__ */ h(this.setItemFp), this.deleteItem = /* @__PURE__ */ h(this.deleteItemFp), this.clear = /* @__PURE__ */ h(this.clearFp);
  }
}
const Je = se({ createRequestId: Be });
function Ht() {
  return new Wt(u.pipe(
    v(),
    y,
    $,
    Je
  ));
}
const jr = /* @__PURE__ */ Ht();
function Kt({ request: s, ...t }) {
  return A((r) => u.pipe(
    s("web_app_request_emoji_status_access", "emoji_status_access_requested", r),
    l.map((n) => n.status)
  ), { ...t, requires: "web_app_request_emoji_status_access", returns: "task" });
}
// @__NO_SIDE_EFFECTS__
function Qt() {
  return Kt(u.pipe(
    v(),
    y,
    $
  ));
}
const zt = /* @__PURE__ */ Qt(), Or = /* @__PURE__ */ h(zt);
function Nt({ request: s, ...t }) {
  return A((r, n) => u.pipe(
    s("web_app_set_emoji_status", ["emoji_status_set", "emoji_status_failed"], {
      params: {
        custom_emoji_id: r,
        duration: (n || {}).duration
      },
      ...n
    }),
    l.chainW((a) => a.event === "emoji_status_failed" ? l.left(new Et(a.payload.error)) : l.right(void 0))
  ), {
    ...t,
    requires: "web_app_set_emoji_status",
    returns: "task"
  });
}
// @__NO_SIDE_EFFECTS__
function Jt() {
  return Nt(u.pipe(
    v(),
    $,
    y
  ));
}
const Yt = /* @__PURE__ */ Jt(), Dr = /* @__PURE__ */ h(Yt);
class Zt {
  constructor({ postEvent: t, isTma: r, version: n }) {
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * A method that tells if an impact occurred. The Telegram app may play the
     * appropriate haptics based on style value passed.
     * @param style - impact style.
     * @since Mini Apps v6.1
     */
    e(this, "impactOccurredFp");
    /**
     * @see impactOccurredFp
     */
    e(this, "impactOccurred");
    /**
     * A method tells that a task or action has succeeded, failed, or produced
     * a warning. The Telegram app may play the appropriate haptics based on type
     * value passed.
     * @param type - notification type.
     * @since Mini Apps v6.1
     */
    e(this, "notificationOccurredFp");
    /**
     * @see notificationOccurredFp
     */
    e(this, "notificationOccurred");
    /**
     * A method tells that the user has changed a selection. The Telegram app may
     * play the appropriate haptics.
     *
     * Do not use this feedback when the user makes or confirms a selection; use
     * it only when the selection changes.
     * @since Mini Apps v6.1
     */
    e(this, "selectionChangedFp");
    /**
     * @see selectionChangedFp
     */
    e(this, "selectionChanged");
    const a = "web_app_trigger_haptic_feedback", p = E({
      requires: a,
      isTma: r,
      version: n,
      returns: "plain"
    });
    this.isSupported = W(a, n), this.impactOccurredFp = p((o) => t(a, { type: "impact", impact_style: o })), this.notificationOccurredFp = p((o) => t(a, { type: "notification", notification_type: o })), this.selectionChangedFp = p(() => t(a, { type: "selection_change" })), this.impactOccurred = /* @__PURE__ */ h(this.impactOccurredFp), this.notificationOccurred = /* @__PURE__ */ h(this.notificationOccurredFp), this.selectionChanged = /* @__PURE__ */ h(this.selectionChangedFp);
  }
}
function Xt() {
  return new Zt(u.pipe(
    v(),
    T,
    y
  ));
}
const Gr = /* @__PURE__ */ Xt();
function es({ postEvent: s, ...t }) {
  return A(() => s("web_app_add_to_home_screen"), { ...t, requires: "web_app_add_to_home_screen", returns: "either" });
}
// @__NO_SIDE_EFFECTS__
function ts() {
  return es(u.pipe(
    v(),
    y,
    T
  ));
}
const ss = /* @__PURE__ */ ts(), Ur = /* @__PURE__ */ h(ss);
function rs({ request: s, ...t }) {
  return A((r) => u.pipe(
    s("web_app_check_home_screen", "home_screen_checked", r),
    l.map((n) => n.status || "unknown")
  ), { ...t, requires: "web_app_check_home_screen", returns: "task" });
}
// @__NO_SIDE_EFFECTS__
function ns() {
  return rs(u.pipe(
    v(),
    y,
    $
  ));
}
const is = /* @__PURE__ */ ns(), Wr = /* @__PURE__ */ h(is);
class os {
  constructor({ retrieveInitData: t }) {
    e(this, "_state", j());
    e(this, "_raw", j());
    /**
     * Complete component state.
     */
    e(this, "state", C(this._state));
    /**
     * @see InitDataType.auth_date
     */
    e(this, "authDate", this.fromState("auth_date"));
    /**
     * @see InitDataType.can_send_after
     */
    e(this, "canSendAfter", this.fromState("can_send_after"));
    /**
     * Date after which it is allowed to call
     * the [answerWebAppQuery](https://core.telegram.org/bots/api#answerwebappquery) method.
     */
    e(this, "canSendAfterDate", C(() => {
      const t = this.authDate(), r = this.canSendAfter();
      return r && t ? new Date(t.getTime() + r * 1e3) : void 0;
    }));
    /**
     * @see InitDataType.chat
     */
    e(this, "chat", this.fromState("chat"));
    /**
     * @see InitDataType.chat_type
     */
    e(this, "chatType", this.fromState("chat_type"));
    /**
     * @see InitDataType.chat_instance
     */
    e(this, "chatInstance", this.fromState("chat_instance"));
    /**
     * @see InitDataType.hash
     */
    e(this, "hash", this.fromState("hash"));
    /**
     * @see InitDataType.query_id
     */
    e(this, "queryId", this.fromState("query_id"));
    /**
     * Raw representation of init data.
     */
    e(this, "raw", C(this._raw));
    /**
     * @see InitDataType.receiver
     */
    e(this, "receiver", this.fromState("receiver"));
    /**
     * @see InitDataType.signature
     */
    e(this, "signature", this.fromState("signature"));
    /**
     * @see InitDataType.start_param
     */
    e(this, "startParam", this.fromState("start_param"));
    /**
     * @see InitDataType.user
     */
    e(this, "user", this.fromState("user"));
    /**
     * Restores the component state.
     */
    e(this, "restoreFp");
    /**
     * @see restoreFp
     */
    e(this, "restore");
    this.restoreFp = () => u.pipe(
      t(),
      b.map(Q.match(() => {
      }, ({ raw: r, obj: n }) => {
        this._state.set(n), this._raw.set(r);
      }))
    ), this.restore = te(this.restoreFp);
  }
  fromState(t) {
    return C(() => {
      const r = this._state();
      return r ? r[t] : void 0;
    });
  }
}
function as() {
  return new os({
    retrieveInitData() {
      return u.pipe(
        b.Do,
        b.bindW("obj", () => u.pipe(
          le(),
          b.map(({ tgWebAppData: s }) => s ? Q.some(s) : Q.none)
        )),
        b.bindW("raw", lt),
        b.map(({ obj: s, raw: t }) => u.pipe(
          Q.Do,
          Q.bind("obj", () => s),
          Q.bind("raw", () => t)
        ))
      );
    }
  });
}
const Hr = /* @__PURE__ */ as();
class ps {
  constructor({ version: t, request: r, isTma: n }) {
    /**
     * Signal indicating if any invoice is currently opened.
     */
    e(this, "isOpened");
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * Opens an invoice using its slug or URL.
     * @param slug - invoice slug.
     * @param options - additional options.
     * @since Mini Apps v6.1
     * @example
     * const status = await invoice.openSlug('kJNFS331');
     */
    e(this, "openSlugFp");
    /**
     * @see openSlugFp
     */
    e(this, "openSlug");
    /**
     * Opens an invoice using its URL.
     * @param url - invoice URL.
     * @param options - additional options.
     * @since Mini Apps v6.1
     * @example
     * const status = await invoice.openUrl('https://t.me/$kJNFS331');
     */
    e(this, "openUrlFp");
    /**
     * @see openUrlFp
     */
    e(this, "openUrl");
    const a = E({
      version: t,
      isTma: n,
      requires: "web_app_open_invoice",
      returns: "task"
    }), p = j(!1), o = () => {
      p.set(!1);
    };
    this.isSupported = W("web_app_open_invoice", t), this.isOpened = C(p), this.openSlugFp = a((i, c) => u.pipe(
      this.isOpened() ? l.left(new Ee("Invoice is already opened")) : l.right(void 0),
      l.chain(() => (p.set(!0), r("web_app_open_invoice", "invoice_closed", {
        ...c,
        params: { slug: i },
        capture: (m) => i === m.slug
      }))),
      l.mapBoth((m) => (o(), m), (m) => (o(), m.status))
    )), this.openUrlFp = a((i, c) => {
      const { hostname: m, pathname: F } = new URL(i, window.location.href);
      if (m !== "t.me")
        return l.left(new U(`Link has unexpected hostname: ${m}`));
      const _ = F.match(/^\/(\$|invoice\/)([A-Za-z0-9\-_=]+)$/);
      return _ ? this.openSlugFp(_[2], c) : l.left(new U(
        'Expected to receive a link with a pathname in format "/invoice/{slug}" or "/${slug}"'
      ));
    }), this.openUrl = /* @__PURE__ */ h(this.openUrlFp), this.openSlug = /* @__PURE__ */ h(this.openSlugFp);
  }
}
function us() {
  return new ps(u.pipe(v(), $, y));
}
const Kr = /* @__PURE__ */ us();
function cs({ postEvent: s, ...t }) {
  return A((r, n = {}) => {
    if (typeof r == "string")
      try {
        r = new URL(r);
      } catch (a) {
        return b.left(new U(`"${r.toString()}" is invalid URL`, a));
      }
    return s("web_app_open_link", {
      url: r.toString(),
      try_browser: n.tryBrowser,
      try_instant_view: n.tryInstantView
    });
  }, { ...t, returns: "either" });
}
// @__NO_SIDE_EFFECTS__
function ls() {
  return cs(u.pipe(v(), T));
}
const hs = /* @__PURE__ */ ls(), Qr = /* @__PURE__ */ h(hs);
function ds({ postEvent: s, version: t, ...r }) {
  return A((n) => {
    const a = n.toString();
    return a.match(/^https:\/\/t.me\/.+/) ? ee("web_app_open_tg_link", V(t)) ? (n = new URL(n), s("web_app_open_tg_link", { path_full: n.pathname + n.search })) : (window.location.href = a, b.right(void 0)) : b.left(new U(`"${a}" is invalid URL`));
  }, { ...r, returns: "either" });
}
// @__NO_SIDE_EFFECTS__
function _s() {
  return ds(u.pipe(
    v(),
    T,
    y
  ));
}
const Ye = /* @__PURE__ */ _s(), zr = /* @__PURE__ */ h(Ye);
function ms({ openTelegramLink: s, ...t }) {
  return A((r, n) => s(
    "https://t.me/share/url?" + new URLSearchParams({ url: r, text: n || "" }).toString().replace(/\+/g, "%20")
  ), { ...t, returns: "either" });
}
// @__NO_SIDE_EFFECTS__
function fs() {
  return ms({
    ...v(),
    openTelegramLink: Ye
  });
}
const bs = /* @__PURE__ */ fs(), Nr = /* @__PURE__ */ h(bs);
function gs(s) {
  let t = !1, r, n;
  return s.available && (t = !0, r = s.access_requested, n = s.access_granted), {
    available: t,
    accessGranted: n || !1,
    accessRequested: r || !1
  };
}
class Fs {
  constructor({
    version: t,
    request: r,
    postEvent: n,
    storage: a,
    isTma: p,
    isPageReload: o
  }) {
    /**
     * Complete location manager state.
     */
    e(this, "state");
    /**
     * Signal indicating whether the location data tracking is currently available.
     */
    e(this, "isAvailable");
    /**
     * Signal indicating whether the user has granted the app permission to track location data.
     */
    e(this, "isAccessGranted");
    /**
     * Signal indicating whether the app has previously requested permission to track location data.
     */
    e(this, "isAccessRequested");
    /**
     * Signal indicating if the component is currently mounted.
     */
    e(this, "isMounted");
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * Opens the location access settings for bots. Useful when you need to request location access
     * from users who haven't granted it yet.
     *
     * Note that this method can be called only in response to user interaction with the Mini App
     * interface (e.g., a click inside the Mini App or on the main button).
     * @since Mini Apps v8.0
     */
    e(this, "openSettingsFp");
    /**
     * @see openSettingsFp
     */
    e(this, "openSettings");
    /**
     * Requests location data.
     * @since Mini Apps v8.0
     * @returns Promise with location data or null it access was not granted.
     */
    e(this, "requestLocationFp");
    /**
     * @see requestLocationFp
     */
    e(this, "requestLocation");
    /**
     * Mounts the component restoring its state.
     * @since Mini Apps v8.0
     */
    e(this, "mountFp");
    /**
     * @see mountFp
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     */
    e(this, "unmount");
    const i = new J({
      initialState: {
        available: !1,
        accessGranted: !1,
        accessRequested: !1
      },
      onChange: a.set
    }), c = new Ae({
      isPageReload: o,
      restoreState: a.get,
      onMounted: i.setState,
      initialState(w) {
        return u.pipe(
          r("web_app_check_location", "location_checked", w),
          l.map(gs)
        );
      }
    }), m = { version: t, requires: "web_app_check_location", isTma: p }, F = E({
      ...m,
      returns: "either"
    }), _ = E({
      ...m,
      returns: "task"
    }), S = E({
      ...m,
      returns: "task",
      isMounted: c.isMounted
    });
    this.isAvailable = i.getter("available"), this.isAccessRequested = i.getter("accessRequested"), this.isAccessGranted = i.getter("accessGranted"), this.isSupported = W("web_app_check_location", t), this.isMounted = c.isMounted, this.state = i.state, this.unmount = c.unmount, this.mountFp = _(c.mount), this.openSettingsFp = F(() => n("web_app_open_location_settings")), this.requestLocationFp = S((w) => u.pipe(
      r("web_app_request_location", "location_requested", w),
      l.map((k) => {
        if (!k.available)
          return i.setState({ available: !1 }), null;
        const { available: I, ...f } = k;
        return f;
      })
    )), this.mount = /* @__PURE__ */ h(this.mountFp), this.openSettings = /* @__PURE__ */ h(this.openSettingsFp), this.requestLocation = /* @__PURE__ */ h(this.requestLocationFp);
  }
}
function Cs() {
  return new Fs(u.pipe(
    v(),
    T,
    y,
    $,
    Y("locationManager")
  ));
}
const Jr = /* @__PURE__ */ Cs();
class ws {
  constructor({ defaults: t, ...r }) {
    //#region Properties.
    /**
     * The button background color.
     */
    e(this, "bgColor");
    /**
     * True if the button has a shining effect.
     */
    e(this, "hasShineEffect");
    /**
     * True if the button is clickable.
     */
    e(this, "isEnabled");
    /**
     * True if the button loader is visible.
     */
    e(this, "isLoaderVisible");
    /**
     * True if the button is visible.
     */
    e(this, "isVisible");
    /**
     * Signal indicating if the component is currently mounted.
     */
    e(this, "isMounted");
    /**
     * The complete button state.
     */
    e(this, "state");
    /**
     * The button displayed text.
     */
    e(this, "text");
    /**
     * The button text color.
     *
     * Note that this value is computed based on the external defaults. For
     * example, if not explicitly set, this value may be equal to one of theme
     * params colors.
     */
    e(this, "textColor");
    /**
     * The ID of custom emoji icon displayed alongside button text.
     * @since Mini Apps v9.5
     */
    e(this, "iconCustomEmojiId");
    //#endregion
    //#region Methods.
    /**
     * Shows the button.
     */
    e(this, "showFp");
    /**
     * @see showFp
     */
    e(this, "show");
    /**
     * Hides the button.
     */
    e(this, "hideFp");
    /**
     * @see hideFp
     */
    e(this, "hide");
    /**
     * Enables the button.
     */
    e(this, "enableFp");
    /**
     * @see enableFp
     */
    e(this, "enable");
    /**
     * Enables the button.
     */
    e(this, "enableShineEffectFp");
    /**
     * @see enableShineEffectFp
     */
    e(this, "enableShineEffect");
    /**
     * Disables the button.
     */
    e(this, "disableFp");
    /**
     * @see disableFp
     */
    e(this, "disable");
    /**
     * Enables the button.
     */
    e(this, "disableShineEffectFp");
    /**
     * @see disableShineEffectFp
     */
    e(this, "disableShineEffect");
    /**
     * Updates the button background color.
     */
    e(this, "setBgColorFp");
    /**
     * @see setBgColorFp
     */
    e(this, "setBgColor");
    /**
     * Updates the button text color.
     */
    e(this, "setTextColorFp");
    /**
     * @see setTextColorFp
     */
    e(this, "setTextColor");
    /**
     * Updates the button text.
     */
    e(this, "setTextFp");
    /**
     * @see setTextFp
     */
    e(this, "setText");
    /**
     * Updates the button custom emoji ID.
     * @since Mini Apps v9.5
     */
    e(this, "setIconCustomEmojiIdFp");
    /**
     * @see setIconCustomEmojiIdFp
     */
    e(this, "setIconCustomEmojiId");
    /**
     * Shows the button loader.
     */
    e(this, "showLoaderFp");
    /**
     * @see showLoaderFp
     */
    e(this, "showLoader");
    /**
     * Hides the button loader.
     */
    e(this, "hideLoaderFp");
    /**
     * @see hideLoaderFp
     */
    e(this, "hideLoader");
    /**
     * Updates the button state.
     * @param state - updates to apply.
     * @example
     * button.setParams({
     *   text: 'Submit',
     *   isEnabled: true,
     *   hasShineEffect: true,
     * });
     */
    e(this, "setParamsFp");
    e(this, "setParams");
    /**
     * Mounts the component restoring its state.
     */
    e(this, "mountFp");
    /**
     * @see mountFp
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     */
    e(this, "unmount");
    /**
     * Adds a new button listener.
     * @param listener - event listener.
     * @param once - should the listener be called only once.
     * @returns A function to remove bound listener.
     * @example
     * const off = button.onClick(() => {
     *   console.log('User clicked the button');
     *   off();
     * });
     */
    e(this, "onClickFp");
    /**
     * @see onClickFp
     */
    e(this, "onClick");
    /**
     * Removes the button click listener.
     * @param listener - event listener.
     * @param once - should the listener be called only once.
     * @example
     * function listener() {
     *   console.log('User clicked the button');
     *   button.offClick(listener);
     * }
     * button.onClick(listener);
     */
    e(this, "offClickFp");
    /**
     * @see offClickFp
     */
    e(this, "offClick");
    const n = new de({
      ...r,
      version: "100",
      initialState: {
        hasShineEffect: !1,
        isEnabled: !0,
        isLoaderVisible: !1,
        isVisible: !1,
        text: "Continue",
        iconCustomEmojiId: ""
      },
      method: "web_app_setup_main_button",
      payload: (o) => ({
        has_shine_effect: o.hasShineEffect,
        is_visible: o.isVisible,
        is_active: o.isEnabled,
        is_progress_visible: o.isLoaderVisible,
        text: o.text,
        color: o.bgColor,
        text_color: o.textColor,
        icon_custom_emoji_id: o.iconCustomEmojiId
      })
    }), a = (o, i) => {
      const c = n.stateGetter(o);
      return C(() => c() || V(i));
    }, p = (o, i) => {
      const c = n.stateGetter(o);
      return C(() => c() || V(i));
    };
    this.bgColor = a("bgColor", t.bgColor), this.textColor = a("textColor", t.textColor), this.hasShineEffect = n.stateGetter("hasShineEffect"), this.isEnabled = n.stateGetter("isEnabled"), this.isLoaderVisible = n.stateGetter("isLoaderVisible"), this.text = n.stateGetter("text"), this.iconCustomEmojiId = p("iconCustomEmojiId", ""), this.isVisible = n.stateGetter("isVisible"), this.isMounted = n.isMounted, this.state = n.state, [this.setBgColor, this.setBgColorFp] = n.stateSetters("bgColor"), [this.setTextColor, this.setTextColorFp] = n.stateSetters("textColor"), [
      [this.disableShineEffect, this.disableShineEffectFp],
      [this.enableShineEffect, this.enableShineEffectFp]
    ] = n.stateBoolSetters("hasShineEffect"), [
      [this.disable, this.disableFp],
      [this.enable, this.enableFp]
    ] = n.stateBoolSetters("isEnabled"), [
      [this.hideLoader, this.hideLoaderFp],
      [this.showLoader, this.showLoaderFp]
    ] = n.stateBoolSetters("isLoaderVisible"), [this.setText, this.setTextFp] = n.stateSetters("text"), [this.setIconCustomEmojiId, this.setIconCustomEmojiIdFp] = n.stateSetters("iconCustomEmojiId"), [[this.hide, this.hideFp], [this.show, this.showFp]] = n.stateBoolSetters("isVisible"), this.setParams = n.setState, this.setParamsFp = n.setStateFp, this.onClick = n.onClick, this.onClickFp = n.onClickFp, this.offClick = n.offClick, this.offClickFp = n.offClickFp, this.mount = n.mount, this.mountFp = n.mountFp, this.unmount = n.unmount;
  }
  //#endregion
}
function Me(s, t) {
  document.documentElement.style.setProperty(s, t);
}
function Ve(s) {
  document.documentElement.style.removeProperty(s);
}
function Ss(s) {
  return u.pipe(
    ft(s),
    b.map((t) => Math.sqrt(
      [0.299, 0.587, 0.114].reduce((r, n, a) => {
        const p = parseInt(t.slice(1 + a * 2, 1 + (a + 1) * 2), 16);
        return r + p * p * n;
      }, 0)
    ) < 120)
  );
}
const Ze = te(Ss);
class vs {
  constructor({
    initialState: t,
    onChange: r,
    offChange: n,
    isTma: a,
    storage: p,
    isPageReload: o
  }) {
    //#region Colors.
    /**
     * @since v6.10
     */
    e(this, "accentTextColor");
    e(this, "bgColor");
    e(this, "buttonColor");
    e(this, "buttonTextColor");
    /**
     * @since v7.10
     */
    e(this, "bottomBarBgColor");
    e(this, "destructiveTextColor");
    /**
     * @since v6.10
     */
    e(this, "headerBgColor");
    e(this, "hintColor");
    e(this, "linkColor");
    e(this, "secondaryBgColor");
    /**
     * @since v6.10
     */
    e(this, "sectionBgColor");
    /**
     * @since v6.10
     */
    e(this, "sectionHeaderTextColor");
    /**
     * @since v7.6
     */
    e(this, "sectionSeparatorColor");
    /**
     * @since v6.10
     */
    e(this, "subtitleTextColor");
    e(this, "textColor");
    //#endregion
    //#region CSS variables.
    e(this, "_isCssVarsBound", j(!1));
    /**
     * True if CSS variables are currently bound.
     */
    e(this, "isCssVarsBound", C(this._isCssVarsBound));
    /**
     * Creates CSS variables connected with the current theme parameters.
     *
     * By default, created CSS variables names are following the pattern "--tg-theme-{name}", where
     * {name} is a theme parameters key name converted from snake case to kebab case.
     *
     * Default variables:
     * - `--tg-theme-bg-color`
     * - `--tg-theme-secondary-text-color`
     *
     * Variables are being automatically updated if theme parameters were changed.
     *
     * @param getCSSVarName - function, returning complete CSS variable name for the specified
     * theme parameters key.
     * @returns Function to stop updating variables.
     * @throws {CSSVarsBoundError} CSS variables are already bound
     * @example Using custom CSS vars generator
     * themeParams.bindCssVars(key => `--my-prefix-${key}`);
     */
    e(this, "bindCssVarsFp");
    /**
     * @see bindCssVarsFp
     */
    e(this, "bindCssVars");
    //#endregion
    //#region Other public signals.
    /**
     * Complete component state.
     */
    e(this, "state");
    /**
     * @returns True if the current color scheme is recognized as dark.
     * This value is calculated based on the current theme's background color.
     */
    e(this, "isDark", C(() => {
      const t = this.bgColor();
      return !t || Ze(t);
    }));
    //#endregion
    //#region Mounting.
    /**
     * Signal indicating if the component is currently mounted.
     */
    e(this, "isMounted");
    /**
     * Mounts the component restoring its state.
     */
    e(this, "mountFp");
    /**
     * @see mountFp
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     */
    e(this, "unmount");
    const i = new J({
      initialState: {},
      onChange: p.set
    }), c = (w) => {
      i.setState(w.theme_params);
    }, m = new ie({
      initialState: () => b.right(V(t)),
      isPageReload: o,
      onMounted(w) {
        i.setState(w), r(c);
      },
      onUnmounted() {
        n(c);
      },
      restoreState: p.get
    }), F = { isTma: a, returns: "either" }, _ = E(F), S = E({
      ...F,
      isMounted: m.isMounted
    });
    this.accentTextColor = i.getter("accent_text_color"), this.bgColor = i.getter("bg_color"), this.buttonColor = i.getter("button_color"), this.buttonTextColor = i.getter("button_text_color"), this.bottomBarBgColor = i.getter("bottom_bar_bg_color"), this.destructiveTextColor = i.getter("destructive_text_color"), this.headerBgColor = i.getter("header_bg_color"), this.hintColor = i.getter("hint_color"), this.linkColor = i.getter("link_color"), this.secondaryBgColor = i.getter("secondary_bg_color"), this.sectionBgColor = i.getter("section_bg_color"), this.sectionHeaderTextColor = i.getter("section_header_text_color"), this.sectionSeparatorColor = i.getter("section_separator_color"), this.subtitleTextColor = i.getter("subtitle_text_color"), this.textColor = i.getter("text_color"), this.state = i.state, this.isMounted = m.isMounted, this.bindCssVarsFp = S((w) => {
      if (this._isCssVarsBound())
        return b.left(new ye());
      w || (w = (f) => `--tg-theme-${mt(f)}`);
      const k = (f) => {
        Object.entries(i.state()).forEach(([g, d]) => {
          d && f(g, d);
        });
      }, I = () => {
        k((f, g) => {
          Me(w(f), g);
        });
      };
      return I(), i.state.sub(I), this._isCssVarsBound.set(!0), b.right(() => {
        k(Ve), i.state.unsub(I), this._isCssVarsBound.set(!1);
      });
    }), this.mountFp = _(m.mount), this.unmount = m.unmount, this.bindCssVars = /* @__PURE__ */ h(this.bindCssVarsFp), this.mount = /* @__PURE__ */ h(this.mountFp);
  }
  //#endregion
}
const Se = /* @__PURE__ */ oe({});
function ks() {
  return new vs({
    ...u.pipe(
      v(),
      Y("themeParams")
    ),
    offChange(s) {
      ne("theme_changed", s);
    },
    onChange(s) {
      z("theme_changed", s);
    },
    initialState: Se
  });
}
const he = /* @__PURE__ */ ks();
// @__NO_SIDE_EFFECTS__
function Xe(s, t, r) {
  return u.pipe(
    /* @__PURE__ */ qe(s, t),
    (n) => ({ ...n, defaults: r })
  );
}
const Yr = /* @__PURE__ */ new ws(
  /* @__PURE__ */ Xe("mainButton", "main_button_pressed", {
    bgColor: C(() => he.buttonColor() || "#2481cc"),
    textColor: C(() => he.buttonTextColor() || "#ffffff"),
    iconCustomEmojiId: C(() => "")
  })
);
class Is {
  constructor({
    storage: t,
    isPageReload: r,
    version: n,
    postEvent: a,
    isTma: p,
    theme: o,
    onVisibilityChanged: i,
    offVisibilityChanged: c
  }) {
    //#region Other properties.
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * True if the current Mini App background color is recognized as dark.
     */
    e(this, "isDark", C(() => {
      const t = this.bgColorRgb();
      return t ? Ze(t) : !1;
    }));
    /**
     * Signal indicating if the mini app is currently active.
     */
    e(this, "isActive");
    /**
     * Complete component state.
     */
    e(this, "state");
    //#endregion
    //#region CSS variables.
    /**
     * True if the CSS variables are currently bound.
     */
    e(this, "isCssVarsBound");
    /**
     * Creates CSS variables connected with the mini app.
     *
     * Default variables:
     * - `--tg-bg-color`
     * - `--tg-header-color`
     * - `--tg-bottom-bar-color`
     *
     * Variables are being automatically updated if theme parameters were changed.
     *
     * @param getCSSVarName - function, returning complete CSS variable name for the specified
     * mini app key.
     * @returns Function to stop updating variables.
     * @example Using no arguments
     * miniApp.bindCssVars();
     * @example Using custom CSS vars generator
     * miniApp.bindCssVars(key => `--my-prefix-${key}`);
     */
    e(this, "bindCssVarsFp");
    e(this, "bindCssVars");
    //#endregion
    //#region Mounting.
    /**
     * Signal indicating if the component is mounted.
     */
    e(this, "isMounted");
    /**
     * Mounts the component.
     *
     * This function restores the component state and is automatically saving it in the local storage
     * if it changed.
     * @since Mini Apps v6.1
     */
    e(this, "mountFp");
    /**
     * @see mount
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     */
    e(this, "unmount");
    //#endregion
    //#region Background color.
    /**
     * The Mini App background color.
     *
     * Represents an RGB color, or theme parameters key, like "bg_color", "secondary_bg_color", etc.
     *
     * Note that using a theme parameters key, background color becomes bound to the current
     * theme parameters, making it automatically being updated whenever theme parameters change.
     * In order to remove this bind, use an explicit RGB color.
     */
    e(this, "bgColor");
    /**
     * RGB representation of the background color.
     *
     * This value requires the Theme Params component to be mounted to extract a valid RGB value
     * of the color key.
     */
    e(this, "bgColorRgb");
    /**
     * Updates the mini app background color.
     * @since Mini Apps v6.1
     */
    e(this, "setBgColorFp");
    /**
     * @see setBgColorFp
     */
    e(this, "setBgColor");
    //#endregion
    //#region Header color.
    /**
     * The Mini App header color.
     */
    e(this, "headerColor");
    /**
     * RGB representation of the header color.
     *
     * This value requires the Theme Params component to be mounted to extract a valid RGB value
     * of the color key.
     */
    e(this, "headerColorRgb");
    /**
     * Updates the mini app header color.
     * @since Mini Apps v6.1
     */
    e(this, "setHeaderColorFp");
    /**
     * @see setHeaderColorFp
     */
    e(this, "setHeaderColor");
    //#endregion
    //#region Bottom bar background color.
    /**
     * The Mini App bottom bar background color.
     */
    e(this, "bottomBarColor");
    /**
     * RGB representation of the bottom bar background color.
     *
     * This value requires the Theme Params component to be mounted to extract a valid RGB value
     * of the color key.
     */
    e(this, "bottomBarColorRgb");
    /**
     * Updates the mini app bottom bar bg color.
     * @since Mini Apps v7.10
     */
    e(this, "setBottomBarColorFp");
    /**
     * @see setBottomBarColorFp
     */
    e(this, "setBottomBarColor");
    //#endregion
    //#region Other methods.
    /**
     * Closes the Mini App.
     * @param returnBack - should the client return to the previous activity.
     */
    e(this, "closeFp");
    /**
     * @see closeFp
     */
    e(this, "close");
    /**
     * Informs the Telegram app that the Mini App is ready to be displayed.
     *
     * It is recommended to call this method as early as possible, as soon as all
     * essential interface elements loaded.
     *
     * Once this method is called, the loading placeholder is hidden and the Mini
     * App shown.
     *
     * If the method is not called, the placeholder will be hidden only when the
     * page was fully loaded.
     */
    e(this, "readyFp");
    /**
     * @see readyFp
     */
    e(this, "ready");
    const m = (d) => {
      S.setState({ isActive: d.is_visible });
    }, F = (d) => {
      [
        [this.headerColor, "web_app_set_header_color"],
        [this.bgColor, "web_app_set_background_color"],
        [this.bottomBarColor, "web_app_set_bottom_bar_color"]
      ].forEach(([B, L]) => {
        const M = B();
        if (!_e(M) && (L !== "web_app_set_header_color" || !["bg_color", "secondary_bg_color"].includes(M))) {
          const P = d[M];
          P && a(L, { color: P });
        }
      });
    }, _ = new ie({
      initialState() {
        return b.right({
          bgColor: "bg_color",
          headerColor: "header_bg_color",
          bottomBarColor: "bottom_bar_bg_color",
          isActive: !0
        });
      },
      isPageReload: r,
      onMounted: (d) => {
        i(m), o.sub(F), S.setState(d);
      },
      onUnmounted() {
        c(m), o.unsub(F);
      },
      restoreState: t.get
    });
    this.isMounted = _.isMounted, this.mountFp = A(() => {
      const d = () => {
      };
      return u.pipe(_.mount(), b.match(d, d));
    }, { isTma: p, returns: "plain" }), this.mount = /* @__PURE__ */ h(this.mountFp), this.unmount = _.unmount;
    const S = new J({
      initialState: {
        bgColor: "bg_color",
        bottomBarColor: "bottom_bar_bg_color",
        headerColor: "bg_color",
        isActive: !1
      },
      onChange: t.set
    });
    this.state = S.state;
    const w = (d) => _e(d) ? d : V(o)[d], k = (d) => C(() => w(d()));
    this.isActive = S.getter("isActive"), this.isSupported = C(() => [
      "web_app_set_header_color",
      "web_app_set_background_color",
      "web_app_set_bottom_bar_color"
    ].some((d) => ee(d, V(n))));
    const I = j(!1);
    this.isCssVarsBound = C(I), this.bindCssVarsFp = A((d) => {
      if (I())
        return b.left(new ye());
      const [B, L] = ke(() => {
        I.set(!1);
      }), M = (P, R) => {
        const G = () => {
          Me(P, R() || null);
        };
        G(), B(R.sub(G), Ve.bind(null, P));
      };
      return d || (d = (P) => `--tg-${Ge(P)}`), M(d("bgColor"), this.bgColorRgb), M(d("bottomBarColor"), this.bottomBarColorRgb), M(d("headerColor"), this.headerColorRgb), I.set(!0), b.right(L);
    }, { isTma: p, returns: "either", isMounted: this.isMounted }), this.bindCssVars = /* @__PURE__ */ h(this.bindCssVarsFp);
    const f = (d) => {
      const B = S.getter(d), L = k(B), M = {
        headerColor: "web_app_set_header_color",
        bgColor: "web_app_set_background_color",
        bottomBarColor: "web_app_set_bottom_bar_color"
      }[d], P = A(
        (R) => {
          if (R === B())
            return b.right(void 0);
          if (M === "web_app_set_header_color" && (R === "bg_color" || R === "secondary_bg_color"))
            return u.pipe(
              a("web_app_set_header_color", { color_key: R }),
              b.map(() => {
                S.setState({ [d]: R });
              })
            );
          const G = w(R);
          return u.pipe(
            G ? a(M, { color: G }) : b.left(new Bt(R)),
            b.map(() => {
              S.setState({ [d]: R });
            })
          );
        },
        {
          isTma: p,
          version: n,
          requires: M,
          isMounted: this.isMounted,
          returns: "either",
          supports: d === "headerColor" ? {
            rgb: {
              method: "web_app_set_header_color",
              param: "color",
              shouldCheck: _e
            }
          } : void 0
        }
      );
      return [B, L, /* @__PURE__ */ h(P), P];
    };
    [
      this.bgColor,
      this.bgColorRgb,
      this.setBgColor,
      this.setBgColorFp
    ] = f("bgColor"), [
      this.headerColor,
      this.headerColorRgb,
      this.setHeaderColor,
      this.setHeaderColorFp
    ] = f("headerColor"), [
      this.bottomBarColor,
      this.bottomBarColorRgb,
      this.setBottomBarColor,
      this.setBottomBarColorFp
    ] = f("bottomBarColor");
    const g = E({ isTma: p, returns: "either" });
    this.closeFp = g((d) => a("web_app_close", { return_back: d })), this.close = /* @__PURE__ */ h(this.closeFp), this.readyFp = g(() => a("web_app_ready")), this.ready = /* @__PURE__ */ h(this.readyFp);
  }
  //#endregion
}
function ys() {
  return new Is({
    ...u.pipe(
      v(),
      T,
      y,
      Y("miniApp")
    ),
    offVisibilityChanged(s) {
      ne("visibility_changed", s);
    },
    onVisibilityChanged(s) {
      z("visibility_changed", s);
    },
    theme: he.state
  });
}
const Es = /* @__PURE__ */ ys();
function qs(s) {
  const t = s.message.trim(), r = (s.title || "").trim(), n = s.buttons || [];
  if (r.length > 64)
    return b.left(new U(`Invalid title: ${r}`));
  if (!t || t.length > 256)
    return b.left(new U(`Invalid message: ${t}`));
  if (n.length > 3)
    return b.left(new U(`Invalid buttons count: ${n.length}`));
  const a = [];
  if (!n.length)
    a.push({ type: "close", id: "" });
  else
    for (let p = 0; p < n.length; p++) {
      const o = n[p], i = o.id || "";
      if (i.length > 64)
        return b.left(new U(`Button with index ${p} has invalid id: ${i}`));
      if (!o.type || o.type === "default" || o.type === "destructive") {
        const c = o.text.trim();
        if (!c || c.length > 64)
          return b.left(new U(`Button with index ${p} has invalid text: ${c}`));
        a.push({ type: o.type, text: c, id: i });
      } else
        a.push({ type: o.type, id: i });
    }
  return b.right({ title: r, message: t, buttons: a });
}
class As {
  constructor({ version: t, isTma: r, request: n }) {
    /**
     * Signal indicating if any popup is currently opened.
     */
    e(this, "isOpened");
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * A method that shows a native popup described by the `params` argument.
     * The promise will be resolved when the popup is closed. Resolved value will have
     * an identifier of the pressed button.
     *
     * If a user clicked outside the popup or clicked the top right popup close
     * button, null will be resolved.
     *
     * @param options - popup parameters.
     * @since Mini Apps v6.2
     * @example
     * fn.pipe(
     *   popup.showFp({
     *     title: 'Confirm action',
     *     message: 'Do you really want to buy this burger?',
     *     buttons: [
     *       { id: 'yes', text: 'Yes' },
     *       { id: 'no', type: 'destructive', text: 'No' },
     *     ],
     *   }),
     *   TE.map(buttonId => {
     *     console.log('User clicked a button with ID', buttonId);
     *   }),
     * );
     */
    e(this, "showFp");
    /**
     * @see showFp
     */
    e(this, "show");
    const a = j(!1), p = () => {
      a.set(!1);
    }, o = E({
      version: t,
      isTma: r,
      requires: "web_app_open_popup",
      returns: "task"
    });
    this.isSupported = W("web_app_open_popup", t), this.isOpened = C(a), this.showFp = o((i) => u.pipe(
      this.isOpened() ? l.left(new Ee("A popup is already opened")) : l.right(void 0),
      l.chainW(() => l.fromEither(qs(i))),
      l.chain((c) => (a.set(!0), n("web_app_open_popup", "popup_closed", {
        ...i,
        params: c
      }))),
      l.mapBoth(
        (c) => (p(), c),
        (c) => (p(), c.button_id)
      )
    )), this.show = /* @__PURE__ */ h(this.showFp);
  }
}
function Bs() {
  return new As(u.pipe(v(), $, y));
}
const Zr = /* @__PURE__ */ Bs();
function xs({ request: s, ...t }) {
  return A((r) => u.pipe(
    s("web_app_request_phone", "phone_requested", r),
    l.map((n) => n.status)
  ), { ...t, requires: "web_app_request_phone", returns: "task" });
}
// @__NO_SIDE_EFFECTS__
function Ms() {
  return xs(u.pipe(
    v(),
    y,
    $
  ));
}
const et = /* @__PURE__ */ Ms(), Xr = /* @__PURE__ */ h(et);
function Vs({
  invokeCustomMethod: s,
  requestPhoneAccess: t,
  ...r
}) {
  const n = (o) => u.pipe(
    s("getRequestedContact", {}, {
      ...o,
      timeout: (o || {}).timeout || 5e3
    }),
    l.chainW((i) => {
      const c = Fe(K(), i);
      if (!c.success)
        return l.left(new ue(i, c.issues));
      if (!c.output)
        return l.right(void 0);
      const m = Fe(
        bt(
          je({
            contact: gt(je({
              user_id: Ke(),
              phone_number: K(),
              first_name: K(),
              last_name: wt(K())
            })),
            auth_date: Ue(
              K(),
              We((F) => new Date(Number(F) * 1e3)),
              He()
            ),
            hash: K()
          })
        ),
        c.output
      );
      return m.success ? l.right({ raw: c.output, parsed: m.output }) : l.left(new ue(c.output, m.issues));
    })
  ), a = (o) => u.pipe(
    n(o),
    l.match(
      // All other errors except validation ones should be ignored. Receiving validation error
      // means that we have some data, but we are unable to parse it properly. So, there is no
      // need to make some more requests further, the problem is local.
      (i) => ue.is(i) ? b.left(i) : b.right(void 0),
      (i) => b.right(i)
    )
  ), p = (o) => ge(
    async (i, c, m) => {
      let F = 50;
      for (; !m.isRejected; ) {
        const _ = await a(m)();
        if (_._tag === "Left")
          return c(_.left);
        if (_.right)
          return i(_.right);
        await new Promise((S) => setTimeout(S, F)), F += 50;
      }
    },
    o
  );
  return A((o) => ge.fn((i) => u.pipe(
    // Try to get the requested contact. Probably, we already requested it before.
    a(i),
    l.chain((c) => c ? l.right(c) : u.pipe(
      t(i),
      l.chainW((m) => m === "sent" ? p(i) : l.left(new Qe("User denied access")))
    ))
  ), o), { ...r, returns: "task", requires: "web_app_request_phone" });
}
// @__NO_SIDE_EFFECTS__
function Ts() {
  return Vs({
    ...u.pipe(v(), xe, y),
    requestPhoneAccess: et
  });
}
function Ls({ requestContact: s, ...t }) {
  return A(
    s,
    { ...t, returns: "task", requires: "web_app_request_phone" }
  );
}
// @__NO_SIDE_EFFECTS__
function $s() {
  return Ls({
    ...u.pipe(v(), y),
    requestContact(s) {
      return u.pipe(
        tt(s),
        l.map((t) => t.parsed)
      );
    }
  });
}
const tt = /* @__PURE__ */ Ts(), en = /* @__PURE__ */ h(tt), Rs = /* @__PURE__ */ $s(), tn = /* @__PURE__ */ h(Rs);
function Ps({ request: s, ...t }) {
  return A((r) => u.pipe(
    s("web_app_request_write_access", "write_access_requested", r),
    l.map((n) => n.status)
  ), { ...t, requires: "web_app_request_write_access", returns: "task" });
}
// @__NO_SIDE_EFFECTS__
function js() {
  return Ps(u.pipe(
    v(),
    y,
    $
  ));
}
const Os = /* @__PURE__ */ js(), sn = /* @__PURE__ */ h(Os);
class Ds {
  constructor({
    version: t,
    onClosed: r,
    onTextReceived: n,
    isTma: a,
    postEvent: p
  }) {
    /**
     * Signal indicating if the scanner is currently opened.
     */
    e(this, "isOpened");
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * Opens the scanner and returns a task which will be completed with the QR content if the
     * passed `capture` function returned true.
     *
     * Task may also be completed with undefined if the scanner was closed.
     * @param options - method options.
     * @returns A promise with QR content presented as string or undefined if the scanner was closed.
     * @since Mini Apps v6.4
     * @example
     * fn.pipe(
     *   qrScanner.captureFp({
     *     capture(scannedQr) {
     *       return scannedQr === 'any expected by me qr';
     *     }
     *   }),
     *   TE.match(
     *     error => {
     *       console.error(error);
     *     },
     *     qr => {
     *       console.log('My QR:'), qr;
     *     }
     *   ),
     * );
     */
    e(this, "captureFp");
    /**
     * @see captureFp
     */
    e(this, "capture");
    /**
     * Closes the scanner.
     * @since Mini Apps v6.4
     */
    e(this, "closeFp");
    /**
     * @see close
     */
    e(this, "close");
    /**
     * Opens the scanner and returns a task which will be completed when the scanner was closed.
     * @param options - method options.
     * @since Mini Apps v6.4
     * @example Without `capture` option
     * if (qrScanner.open.isAvailable()) {
     *   const qr = await qrScanner.open({ text: 'Scan any QR' });
     * }
     * @example
     * fn.pipe(
     *   qrScanner.openFp({
     *     onCaptured(scannedQr) {
     *       if (scannedQr === 'any expected by me qr') {
     *         qrScanner.close();
     *       }
     *     }
     *   }),
     *   TE.match(
     *     error => {
     *       console.error(error);
     *     },
     *     () => {
     *       console.log('The scanner was closed');
     *     }
     *   ),
     * );
     */
    e(this, "openFp");
    /**
     * @see openFp
     */
    e(this, "open");
    const o = { version: t, requires: "web_app_open_scan_qr_popup", isTma: a }, i = E({ ...o, returns: "either" }), c = E({ ...o, returns: "task" }), m = j(!1), F = () => {
      m.set(!1);
    };
    this.isSupported = W("web_app_open_scan_qr_popup", t), this.isOpened = C(m), this.captureFp = c((_) => {
      let S;
      return u.pipe(
        this.openFp({
          ..._,
          onCaptured: (w) => {
            _.capture(w) && (S = w, this.close());
          }
        }),
        l.map(() => S)
      );
    }), this.closeFp = i(() => u.pipe(p("web_app_close_scan_qr_popup"), b.map(F))), this.openFp = c((_) => u.pipe(
      m() ? l.left(new Ee("The QR Scanner is already opened")) : async () => p("web_app_open_scan_qr_popup", { text: _.text }),
      l.chainW(() => {
        m.set(!0);
        const [S, w] = ke(), k = (I) => (w(), m.set(!1), I);
        return u.pipe(
          ge((I) => {
            S(
              // The scanner was closed externally.
              r(I),
              // The scanner was closed internally.
              m.sub((f) => {
                f || I();
              }),
              n(_.onCaptured)
            );
          }, _),
          l.mapBoth(k, k)
        );
      })
    )), this.open = /* @__PURE__ */ h(this.openFp), this.capture = /* @__PURE__ */ h(this.captureFp), this.close = /* @__PURE__ */ h(this.closeFp);
  }
}
function Gs() {
  return new Ds({
    ...u.pipe(v(), T, y),
    onClosed(s) {
      return z("scan_qr_popup_closed", s);
    },
    onTextReceived(s) {
      return z("qr_text_received", (t) => {
        s(t.data);
      });
    }
  });
}
const rn = /* @__PURE__ */ Gs();
class Us {
  constructor({ defaults: t, ...r }) {
    //#region Properties.
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * The button position relative to the main one.
     */
    e(this, "position");
    /**
     * The button background color.
     */
    e(this, "bgColor");
    /**
     * True if the button has a shining effect.
     */
    e(this, "hasShineEffect");
    /**
     * True if the button is clickable.
     */
    e(this, "isEnabled");
    /**
     * True if the button loader is visible.
     */
    e(this, "isLoaderVisible");
    /**
     * True if the button is visible.
     */
    e(this, "isVisible");
    /**
     * Signal indicating if the component is currently mounted.
     */
    e(this, "isMounted");
    /**
     * The complete button state.
     */
    e(this, "state");
    /**
     * The button displayed text.
     */
    e(this, "text");
    /**
     * The button text color.
     *
     * Note that this value is computed based on the external defaults. For
     * example, if not explicitly set, this value may be equal to one of theme
     * params colors.
     */
    e(this, "textColor");
    /**
     * The ID of custom emoji icon displayed alongside button text.
     * @since Mini Apps v9.5
     */
    e(this, "iconCustomEmojiId");
    //#endregion
    //#region Methods.
    /**
     * Shows the button.
     * @since Mini Apps v7.10
     */
    e(this, "showFp");
    /**
     * @see showFp
     */
    e(this, "show");
    /**
     * Hides the button.
     * @since Mini Apps v7.10
     */
    e(this, "hideFp");
    /**
     * @see hideFp
     */
    e(this, "hide");
    /**
     * Enables the button.
     * @since Mini Apps v7.10
     */
    e(this, "enableFp");
    /**
     * @see enableFp
     */
    e(this, "enable");
    /**
     * Enables the button.
     * @since Mini Apps v7.10
     */
    e(this, "enableShineEffectFp");
    /**
     * @see enableShineEffectFp
     */
    e(this, "enableShineEffect");
    /**
     * Disables the button.
     * @since Mini Apps v7.10
     */
    e(this, "disableFp");
    /**
     * @see disableFp
     */
    e(this, "disable");
    /**
     * Enables the button.
     * @since Mini Apps v7.10
     */
    e(this, "disableShineEffectFp");
    /**
     * @see disableShineEffectFp
     */
    e(this, "disableShineEffect");
    /**
     * Updates the button background color.
     * @since Mini Apps v7.10
     */
    e(this, "setBgColorFp");
    /**
     * @see setBgColorFp
     */
    e(this, "setBgColor");
    /**
     * Updates the button text color.
     * @since Mini Apps v7.10
     */
    e(this, "setTextColorFp");
    /**
     * @see setTextColorFp
     */
    e(this, "setTextColor");
    /**
     * Updates the button text.
     * @since Mini Apps v7.10
     */
    e(this, "setTextFp");
    /**
     * @see setTextFp
     */
    e(this, "setText");
    /**
     * Updates the button position.
     * @since Mini Apps v7.10
     */
    e(this, "setPositionFp");
    /**
     * @see setPositionFp
     */
    e(this, "setPosition");
    /**
     * Updates the button custom emoji Id.
     * @since Mini Apps v9.5
     */
    e(this, "setIconCustomEmojiIdFp");
    /**
     * @see setIconCustomEmojiIdFp
     */
    e(this, "setIconCustomEmojiId");
    /**
     * Shows the button loader.
     * @since Mini Apps v7.10
     */
    e(this, "showLoaderFp");
    /**
     * @see showLoaderFp
     */
    e(this, "showLoader");
    /**
     * Hides the button loader.
     * @since Mini Apps v7.10
     */
    e(this, "hideLoaderFp");
    /**
     * @see hideLoaderFp
     */
    e(this, "hideLoader");
    /**
     * Updates the button state.
     * @param state - updates to apply.
     * @since Mini Apps v7.10
     * @example
     * button.setParams({
     *   text: 'Submit',
     *   isEnabled: true,
     *   hasShineEffect: true,
     * });
     */
    e(this, "setParamsFp");
    /**
     * @see setParamsFp
     */
    e(this, "setParams");
    /**
     * Mounts the component restoring its state.
     * @since Mini Apps v7.10
     */
    e(this, "mountFp");
    /**
     * @see mountFp
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     */
    e(this, "unmount");
    /**
     * Adds a new button listener.
     * @param listener - event listener.
     * @param once - should the listener be called only once.
     * @returns A function to remove bound listener.
     * @since Mini Apps v7.10
     * @example
     * const off = button.onClick(() => {
     *   console.log('User clicked the button');
     *   off();
     * });
     */
    e(this, "onClickFp");
    /**
     * @see onClick
     */
    e(this, "onClick");
    /**
     * Removes the button click listener.
     * @param listener - event listener.
     * @param once - should the listener be called only once.
     * @since Mini Apps v7.10
     * @example
     * function listener() {
     *   console.log('User clicked the button');
     *   button.offClick(listener);
     * }
     * button.onClick(listener);
     */
    e(this, "offClickFp");
    /**
     * @see offClick
     */
    e(this, "offClick");
    const n = new de({
      ...r,
      initialState: {
        hasShineEffect: !1,
        isEnabled: !0,
        isLoaderVisible: !1,
        isVisible: !1,
        text: "Cancel",
        position: "left",
        iconCustomEmojiId: ""
      },
      method: "web_app_setup_secondary_button",
      payload: (o) => ({
        has_shine_effect: o.hasShineEffect,
        is_visible: o.isVisible,
        is_active: o.isEnabled,
        is_progress_visible: o.isLoaderVisible,
        text: o.text,
        color: o.bgColor,
        text_color: o.textColor,
        position: o.position,
        icon_custom_emoji_id: o.iconCustomEmojiId
      })
    }), a = (o, i) => {
      const c = n.stateGetter(o);
      return C(() => c() || V(i));
    }, p = (o, i) => {
      const c = n.stateGetter(o);
      return C(() => c() || V(i));
    };
    this.isSupported = W("web_app_setup_secondary_button", r.version), this.bgColor = a("bgColor", t.bgColor), this.textColor = a("textColor", t.textColor), this.position = n.stateGetter("position"), this.hasShineEffect = n.stateGetter("hasShineEffect"), this.isEnabled = n.stateGetter("isEnabled"), this.isLoaderVisible = n.stateGetter("isLoaderVisible"), this.text = n.stateGetter("text"), this.iconCustomEmojiId = p("iconCustomEmojiId", ""), this.isVisible = n.stateGetter("isVisible"), this.isMounted = n.isMounted, this.state = n.state, [this.setPosition, this.setPositionFp] = n.stateSetters("position"), [this.setBgColor, this.setBgColorFp] = n.stateSetters("bgColor"), [this.setTextColor, this.setTextColorFp] = n.stateSetters("textColor"), [
      [this.disableShineEffect, this.disableShineEffectFp],
      [this.enableShineEffect, this.enableShineEffectFp]
    ] = n.stateBoolSetters("hasShineEffect"), [
      [this.disable, this.disableFp],
      [this.enable, this.enableFp]
    ] = n.stateBoolSetters("isEnabled"), [
      [this.hideLoader, this.hideLoaderFp],
      [this.showLoader, this.showLoaderFp]
    ] = n.stateBoolSetters("isLoaderVisible"), [this.setText, this.setTextFp] = n.stateSetters("text"), [this.setIconCustomEmojiId, this.setIconCustomEmojiIdFp] = n.stateSetters("iconCustomEmojiId"), [[this.hide, this.hideFp], [this.show, this.showFp]] = n.stateBoolSetters("isVisible"), this.setParams = n.setState, this.setParamsFp = n.setStateFp, this.onClick = n.onClick, this.onClickFp = n.onClickFp, this.offClick = n.offClick, this.offClickFp = n.offClickFp, this.mount = n.mount, this.mountFp = n.mountFp, this.unmount = n.unmount;
  }
  //#endregion
}
function Ws() {
  return new Us(
    /* @__PURE__ */ Xe("secondaryButton", "secondary_button_pressed", {
      bgColor: C(() => Es.bottomBarColorRgb() || "#000000"),
      textColor: C(() => he.buttonColor() || "#2481cc"),
      iconCustomEmojiId: C(() => "")
    })
  );
}
const nn = /* @__PURE__ */ Ws();
class Hs {
  constructor({ isTma: t, request: r, version: n, createRequestId: a }) {
    /**
      * Retrieves an item using its key.
      * @since Mini Apps v9.0
      */
    e(this, "getItemFp");
    /**
     * @see getItemFp
     */
    e(this, "getItem");
    /**
     * Restores an item from the storage.
     * @since Mini Apps v9.0
     */
    e(this, "restoreItemFp");
    /**
     * @see restoreItemFp
     */
    e(this, "restoreItem");
    /**
      * Sets a new item in the storage.
      * @since Mini Apps v9.0
      */
    e(this, "setItemFp");
    /**
     * @see setItemFp
     */
    e(this, "setItem");
    /**
      * Removes a key from the storage.
      * @since Mini Apps v9.0
      */
    e(this, "deleteItemFp");
    /**
     * @see deleteItemFp
     */
    e(this, "deleteItem");
    /**
      * Removes all keys from the storage.
      * @since Mini Apps v9.0
      */
    e(this, "clearFp");
    /**
     * @see clearFp
     */
    e(this, "clear");
    const p = E({
      version: n,
      requires: "web_app_secure_storage_get_key",
      isTma: t,
      returns: "task"
    }), o = (i, c, m) => {
      const F = a();
      return u.pipe(
        r(i, ["secure_storage_failed", c], {
          params: { ...m, req_id: F },
          capture: (_) => "payload" in _ ? _.payload.req_id === F : !0
        }),
        l.chain((_) => _.event === "secure_storage_failed" ? l.left(new kt(_.payload.error || "UNKNOWN_ERROR")) : l.right(_.payload))
      );
    };
    this.getItemFp = p((i) => u.pipe(
      o("web_app_secure_storage_get_key", "secure_storage_key_received", { key: i }),
      l.map((c) => ({
        value: c.value,
        canRestore: !!c.can_restore
      }))
    )), this.setItemFp = p((i, c) => u.pipe(
      o("web_app_secure_storage_save_key", "secure_storage_key_saved", { key: i, value: c }),
      l.map(() => {
      })
    )), this.deleteItemFp = p((i) => this.setItemFp(i, null)), this.clearFp = p(() => u.pipe(
      o("web_app_secure_storage_clear", "secure_storage_cleared", {}),
      l.map(() => {
      })
    )), this.restoreItemFp = p((i) => u.pipe(
      o("web_app_secure_storage_restore_key", "secure_storage_key_restored", { key: i }),
      l.map((c) => c.value)
    )), this.getItem = /* @__PURE__ */ h(this.getItemFp), this.setItem = /* @__PURE__ */ h(this.setItemFp), this.deleteItem = /* @__PURE__ */ h(this.deleteItemFp), this.clear = /* @__PURE__ */ h(this.clearFp), this.restoreItem = /* @__PURE__ */ h(this.restoreItemFp);
  }
}
function Ks() {
  return new Hs(u.pipe(
    v(),
    y,
    $,
    Je
  ));
}
const on = /* @__PURE__ */ Ks();
class Qs {
  constructor(t) {
    /**
     * Signal indicating if the component is currently visible.
     */
    e(this, "isVisible");
    /**
     * Signal indicating if the component is currently mounted.
     */
    e(this, "isMounted");
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * Hides the button.
     * @since Mini Apps v6.10
     */
    e(this, "hideFp");
    /**
     * @see hideFp
     */
    e(this, "hide");
    /**
     * Shows the button.
     * @since Mini Apps v6.10
     */
    e(this, "showFp");
    /**
     * @see showFp
     */
    e(this, "show");
    /**
     * Adds a new button listener.
     * @param listener - event listener.
     * @param once - should the listener be called only once.
     * @returns A function to remove bound listener.
     * @since Mini Apps v6.10
     * @example
     * const off = button.onClick(() => {
     *   console.log('User clicked the button');
     *   off();
     * });
     */
    e(this, "onClickFp");
    /**
     * @see onClickFp
     */
    e(this, "onClick");
    /**
     * Removes the button click listener.
     * @param listener - event listener.
     * @param once - should the listener be called only once.
     * @since Mini Apps v6.10
     * @example
     * function listener() {
     *   console.log('User clicked the button');
     *   button.offClick(listener);
     * }
     * button.onClick(listener);
     */
    e(this, "offClickFp");
    /**
     * @see offClickFp
     */
    e(this, "offClick");
    /**
     * Mounts the component restoring its state.
     * @since Mini Apps v6.10
     */
    e(this, "mountFp");
    /**
     * @see mountFp
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     *
     * Note that this function does not remove listeners added via the `onClick`
     * function, so you have to remove them on your own.
     * @see onClick
     */
    e(this, "unmount");
    const r = new de({
      ...t,
      method: "web_app_setup_settings_button",
      payload: (n) => ({ is_visible: n.isVisible }),
      initialState: { isVisible: !1 }
    });
    this.isVisible = r.stateGetter("isVisible"), this.isMounted = r.isMounted, this.isSupported = r.isSupported, [[this.hide, this.hideFp], [this.show, this.showFp]] = r.stateBoolSetters("isVisible"), this.onClick = r.onClick, this.onClickFp = r.onClickFp, this.offClick = r.offClick, this.offClickFp = r.offClickFp, this.mount = r.mount, this.mountFp = r.mountFp, this.unmount = r.unmount;
  }
}
const an = /* @__PURE__ */ new Qs(
  /* @__PURE__ */ qe("settingsButton", "settings_button_pressed")
);
class zs {
  constructor({ postEvent: t, storage: r, isTma: n, isPageReload: a, version: p }) {
    /**
     * Signal indicating if the component is supported.
     */
    e(this, "isSupported");
    /**
     * Signal indicating if vertical swipes are enabled.
     */
    e(this, "isVerticalEnabled");
    /**
     * Signal indicating if the component is currently mounted.
     */
    e(this, "isMounted");
    /**
     * Mounts the component restoring its state.
     * @since Mini Apps v7.7
     */
    e(this, "mountFp");
    /**
     * @see mountFp
     */
    e(this, "mount");
    /**
     * Unmounts the component.
     */
    e(this, "unmount");
    /**
     * Disables the closing confirmation dialog.
     * @since Mini Apps v7.7
     */
    e(this, "disableVerticalFp");
    /**
     * @see disableVerticalFp
     */
    e(this, "disableVertical");
    /**
     * Enables the closing confirmation dialog.
     * @since Mini Apps v7.7
     */
    e(this, "enableVerticalFp");
    /**
     * @see enableVerticalFp
     */
    e(this, "enableVertical");
    const o = { isVerticalEnabled: !0 }, i = new J({
      initialState: o,
      onChange(w) {
        r.set(w);
      }
    }), c = new ie({
      initialState: o,
      isPageReload: a,
      onMounted: i.setState,
      restoreState: r.get
    }), m = { requires: "web_app_setup_swipe_behavior", isTma: n, version: p }, F = E({
      ...m,
      returns: "plain"
    }), _ = E({
      ...m,
      isMounted: c.isMounted,
      returns: "either"
    }), S = (w) => {
      const k = { isVerticalEnabled: w };
      return i.hasDiff(k) ? u.pipe(
        t("web_app_setup_swipe_behavior", { allow_vertical_swipe: w }),
        b.map(() => {
          i.setState(k);
        })
      ) : b.right(void 0);
    };
    this.isSupported = W("web_app_setup_swipe_behavior", p), this.isVerticalEnabled = i.getter("isVerticalEnabled"), this.isMounted = c.isMounted, this.disableVerticalFp = _(() => S(!1)), this.enableVerticalFp = _(() => S(!0)), this.mountFp = F(() => {
      const w = () => {
      };
      return u.pipe(c.mount(), b.match(w, w));
    }), this.unmount = c.unmount, this.disableVertical = /* @__PURE__ */ h(this.disableVerticalFp), this.enableVertical = /* @__PURE__ */ h(this.enableVerticalFp), this.mount = /* @__PURE__ */ h(this.mountFp);
  }
}
function Ns() {
  return new zs(u.pipe(
    v(),
    T,
    y,
    Y("swipeBehavior")
  ));
}
const pn = /* @__PURE__ */ Ns();
async function un(s) {
  try {
    const { clipboard: r } = navigator;
    if (r)
      return await r.writeText(s);
  } catch {
  }
  const t = document.createElement("textarea");
  t.value = s, t.style.top = "0", t.style.left = "0", t.style.position = "fixed", document.body.appendChild(t), t.focus(), t.select();
  try {
    document.execCommand("copy");
  } finally {
    document.body.removeChild(t);
  }
}
function Js({ request: s, ...t }) {
  return A((r, n, a) => u.pipe(
    s(
      "web_app_request_file_download",
      "file_download_requested",
      { ...a, params: { url: r, file_name: n } }
    ),
    l.chain((p) => p.status === "downloading" ? l.right(void 0) : l.left(new Qe("User denied the action")))
  ), { ...t, requires: "web_app_request_file_download", returns: "task" });
}
// @__NO_SIDE_EFFECTS__
function Ys() {
  return Js(u.pipe(
    v(),
    $,
    y
  ));
}
const Zs = /* @__PURE__ */ Ys(), cn = /* @__PURE__ */ h(Zs);
function Xs({ invokeCustomMethod: s, ...t }) {
  return A((r) => u.pipe(
    s("getCurrentTime", {}, r),
    l.chain((n) => {
      const a = Fe(
        Ue(Ke(), St(), We((p) => new Date(p * 1e3)), He()),
        n
      );
      return a.success ? l.right(a.output) : l.left(new ue(n, a.issues));
    })
  ), { ...t, requires: "web_app_invoke_custom_method", returns: "task" });
}
// @__NO_SIDE_EFFECTS__
function er() {
  return Xs(u.pipe(
    v(),
    xe,
    y
  ));
}
const tr = /* @__PURE__ */ er(), ln = /* @__PURE__ */ h(tr);
function sr({ postEvent: s, ...t }) {
  return A(() => s("web_app_hide_keyboard"), { ...t, returns: "either", requires: "web_app_hide_keyboard" });
}
// @__NO_SIDE_EFFECTS__
function rr() {
  return sr(u.pipe(
    v(),
    T,
    y
  ));
}
const nr = /* @__PURE__ */ rr(), hn = /* @__PURE__ */ h(nr);
function ir({ request: s, createRequestId: t, ...r }) {
  return A((n) => {
    const a = t();
    return u.pipe(
      s("web_app_read_text_from_clipboard", "clipboard_text_received", {
        ...n,
        params: { req_id: a },
        capture: ht(a)
      }),
      l.map(({ data: p = null }) => p)
    );
  }, { ...r, requires: "web_app_read_text_from_clipboard", returns: "task" });
}
// @__NO_SIDE_EFFECTS__
function or() {
  return ir({
    ...u.pipe(
      v(),
      y,
      $
    ),
    createRequestId: Be
  });
}
const ar = /* @__PURE__ */ or(), dn = /* @__PURE__ */ h(ar);
function pr(s) {
  const t = {}, r = s.match(/Telegram-Android(?:\/([^ ]+))?(?: (\([^)]+\))?|$)/);
  if (r) {
    const [, n, a] = r;
    n && (t.appVersion = n), a && a.slice(1, a.length - 1).split(";").forEach((p) => {
      const [o, i] = p.trim().split(" ");
      if (o === "Android")
        t.androidVersion = i;
      else if (o === "SDK") {
        const c = parseInt(i, 10);
        c && (t.sdkVersion = c);
      } else i ? (t.manufacturer = o, t.model = i) : t.performanceClass = o;
    });
  }
  return t;
}
function _n() {
  return pr(navigator.userAgent);
}
function ur({ postEvent: s, ...t }) {
  return A((r) => {
    const { size: n } = new Blob([r]);
    return !n || n > 4096 ? b.left(
      new U(n ? "Maximum size of data to send is 4096 bytes" : "Attempted to send empty data")
    ) : s("web_app_data_send", { data: r });
  }, { ...t, returns: "either" });
}
// @__NO_SIDE_EFFECTS__
function cr() {
  return ur(u.pipe(v(), T));
}
const lr = /* @__PURE__ */ cr(), mn = /* @__PURE__ */ h(lr);
function hr({ request: s, ...t }) {
  return A((r, n) => u.pipe(
    s(
      "web_app_send_prepared_message",
      ["prepared_message_failed", "prepared_message_sent"],
      {
        ...n,
        params: { id: r }
      }
    ),
    l.chain((a) => a.event === "prepared_message_failed" ? l.left(new At(a.payload.error)) : l.right(void 0))
  ), { ...t, requires: "web_app_send_prepared_message", returns: "task" });
}
// @__NO_SIDE_EFFECTS__
function dr() {
  return hr(u.pipe(
    v(),
    $,
    y
  ));
}
const _r = /* @__PURE__ */ dr(), fn = /* @__PURE__ */ h(_r);
function mr({ postEvent: s, ...t }) {
  return A((r, n = {}) => s("web_app_share_to_story", {
    text: n.text,
    media_url: r,
    widget_link: n.widgetLink
  }), { ...t, requires: "web_app_share_to_story", returns: "either" });
}
// @__NO_SIDE_EFFECTS__
function fr() {
  return mr(u.pipe(
    v(),
    T,
    y
  ));
}
const br = /* @__PURE__ */ fr(), bn = /* @__PURE__ */ h(br), ve = /* @__PURE__ */ oe(!1);
function gr({ isInlineMode: s, postEvent: t, ...r }) {
  return A((n, a) => t("web_app_switch_inline_query", {
    query: n,
    chat_types: a || []
  }), {
    ...r,
    requires: {
      every: ["web_app_switch_inline_query", () => V(s) ? void 0 : "The application must be launched in the inline mode"]
    },
    returns: "either"
  });
}
// @__NO_SIDE_EFFECTS__
function Fr() {
  return gr({
    ...u.pipe(
      v(),
      T,
      y
    ),
    isInlineMode: ve
  });
}
const Cr = /* @__PURE__ */ Fr(), gn = /* @__PURE__ */ h(Cr);
class wr {
  constructor({
    storage: t,
    isPageReload: r,
    onContentSafeAreaInsetsChanged: n,
    onSafeAreaInsetsChanged: a,
    onViewportChanged: p,
    onFullscreenChanged: o,
    offContentSafeAreaInsetsChanged: i,
    offFullscreenChanged: c,
    offSafeAreaInsetsChanged: m,
    offViewportChanged: F,
    request: _,
    isViewportStable: S,
    isFullscreen: w,
    isTma: k,
    version: I,
    postEvent: f
  }) {
    //#region Other properties.
    /**
     * Complete component state.
     */
    e(this, "state");
    /**
     * Signal containing the current height of the **visible area** of the Mini App.
     *
     * The application can display just the top part of the Mini App, with its
     * lower part remaining outside the screen area. From this position, the user
     * can "pull" the Mini App to its maximum height, while the bot can do the same
     * by calling `expand` method. As the position of the Mini App changes, the
     * current height value of the visible area will be updated  in real time.
     *
     * Please note that the refresh rate of this value is not sufficient to
     * smoothly follow the lower border of the window. It should not be used to pin
     * interface elements to the bottom of the visible area. It's more appropriate
     * to use the value of the `stableHeight` field for this purpose.
     *
     * @see stableHeight
     */
    e(this, "height");
    /**
     * Signal containing the height of the visible area of the Mini App in its last stable state.
     *
     * The application can display just the top part of the Mini App, with its
     * lower part remaining outside the screen area. From this position, the user
     * can "pull" the Mini App to its maximum height, while the application can do
     * the same by calling `expand` method.
     *
     * Unlike the value of `height`, the value of `stableHeight` does not change as
     * the position of the Mini App changes with user gestures or during
     * animations. The value of `stableHeight` will be updated after all gestures
     * and animations are completed and the Mini App reaches its final size.
     *
     * @see height
     */
    e(this, "stableHeight");
    /**
     * Signal containing the currently visible area width.
     */
    e(this, "width");
    /**
     * Signal indicating if the Mini App is expanded to the maximum available height. Otherwise,
     * if the Mini App occupies part of the screen and can be expanded to the full
     * height using the `expand` method.
     */
    e(this, "isExpanded");
    /**
     * Signal indicating if the current viewport height is stable and is not going to change in
     * the next moment.
     */
    e(this, "isStable", C(() => this.height() === this.stableHeight()));
    //#endregion
    //#region Content safe area insets.
    /**
     * Signal containing content safe area insets.
     */
    e(this, "contentSafeAreaInsets");
    /**
     * Signal containing top content safe area inset.
     */
    e(this, "contentSafeAreaInsetTop");
    /**
     * Signal containing left content safe area inset.
     */
    e(this, "contentSafeAreaInsetLeft");
    /**
     * Signal containing right content safe area inset.
     */
    e(this, "contentSafeAreaInsetRight");
    /**
     * Signal containing bottom content safe area inset.
     */
    e(this, "contentSafeAreaInsetBottom");
    //#endregion
    //#region Safe area insets.
    /**
     * Signal containing safe area insets.
     */
    e(this, "safeAreaInsets");
    /**
     * Signal containing top safe area inset.
     */
    e(this, "safeAreaInsetTop");
    /**
     * Signal containing left safe area inset.
     */
    e(this, "safeAreaInsetLeft");
    /**
     * Signal containing right safe area inset.
     */
    e(this, "safeAreaInsetRight");
    /**
     * Signal containing bottom safe area inset.
     */
    e(this, "safeAreaInsetBottom");
    //#endregion
    //#region Fullscreen.
    /**
     * Signal indicating if the viewport is currently in fullscreen mode.
     */
    e(this, "isFullscreen");
    /**
     * Requests fullscreen mode for the mini application.
     * @since Mini Apps v8.0
     */
    e(this, "requestFullscreenFp");
    /**
     * @see requestFullscreenFp
     */
    e(this, "requestFullscreen");
    /**
     * Exits mini application from the fullscreen mode.
     * @since Mini Apps v8.0
     */
    e(this, "exitFullscreenFp");
    /**
     * @see exitFullscreenFp
     */
    e(this, "exitFullscreen");
    //#endregion
    //#region CSS Vars.
    /**
     * Signal indicating if CSS variables are bound.
     */
    e(this, "isCssVarsBound");
    /**
     * Creates CSS variables connected with the current viewport.
     *
     * By default, created CSS variables names are following the pattern "--tg-theme-{name}", where
     * {name} is a viewport property name converted from camel case to kebab case.
     *
     * Default variables:
     * - `--tg-viewport-height`
     * - `--tg-viewport-width`
     * - `--tg-viewport-stable-height`
     * - `--tg-viewport-content-safe-area-inset-top`
     * - `--tg-viewport-content-safe-area-inset-bottom`
     * - `--tg-viewport-content-safe-area-inset-left`
     * - `--tg-viewport-content-safe-area-inset-right`
     * - `--tg-viewport-safe-area-inset-top`
     * - `--tg-viewport-safe-area-inset-bottom`
     * - `--tg-viewport-safe-area-inset-left`
     * - `--tg-viewport-safe-area-inset-right`
     *
     * Variables are being automatically updated if the viewport was changed.
     *
     * @param getCSSVarName - function, returning computed complete CSS variable name. The CSS
     * variable will only be defined if the function returned non-empty string value.
     * @returns Function to stop updating variables.
     * @example Using no arguments
     * bindCssVarsFp();
     * @example Using custom CSS vars generator
     * bindCssVarsFp(key => `--my-prefix-${key}`);
     */
    e(this, "bindCssVarsFp");
    /**
     * @see bindCssVarsFp
     */
    e(this, "bindCssVars");
    //#endregion
    //#region Mount.
    /**
     * Signal indicating if the component is currently mounted.
     */
    e(this, "isMounted");
    /**
     * Mounts the component.
     */
    e(this, "mountFp");
    /**
     * @see mountFp
     */
    e(this, "mount");
    //#endregion
    //#region Other methods.
    /**
     * A method that expands the Mini App to the maximum available height. To find
     * out if the Mini App is expanded to the maximum height, refer to the value of
     * the `isExpanded`.
     */
    e(this, "expandFp");
    /**
     * @see expandFp
     */
    e(this, "expand");
    const g = { top: 0, right: 0, left: 0, bottom: 0 }, d = new J({
      initialState: {
        contentSafeAreaInsets: g,
        height: 0,
        isExpanded: !1,
        isFullscreen: !1,
        safeAreaInsets: g,
        stableHeight: 0,
        width: 0
      },
      onChange: t.set
    }), B = (q) => {
      d.setState({
        isExpanded: q.is_expanded,
        height: q.height,
        width: q.width,
        stableHeight: q.is_state_stable ? q.height : void 0
      });
    }, L = (q) => {
      d.setState({ isFullscreen: q.is_fullscreen });
    }, M = (q) => {
      d.setState({ safeAreaInsets: q });
    }, P = (q) => {
      d.setState({ contentSafeAreaInsets: q });
    }, R = new Ae({
      initialState(q) {
        const Z = (H) => () => {
          const [N, O] = H === "safe-area" ? ["web_app_request_safe_area", "safe_area_changed"] : ["web_app_request_content_safe_area", "content_safe_area_changed"];
          return ee(N, V(I)) ? _(N, O, q) : l.right({ top: 0, left: 0, right: 0, bottom: 0 });
        }, x = (H) => () => typeof H == "boolean" ? l.right(H) : l.fromEither(H());
        return u.pipe(
          l.Do,
          l.bindW("safeAreaInsets", Z("safe-area")),
          l.bindW("contentSafeAreaInsets", Z("content-safe-area")),
          l.bindW("isFullscreen", x(w)),
          l.bindW("isViewportStable", x(S)),
          l.chainW(({ isViewportStable: H, ...N }) => H ? l.right({
            ...N,
            height: window.innerHeight,
            isExpanded: !0,
            stableHeight: window.innerHeight,
            width: window.innerWidth
          }) : u.pipe(
            _("web_app_request_viewport", "viewport_changed", q),
            l.map((O) => ({
              ...N,
              height: O.height,
              isExpanded: O.is_expanded,
              stableHeight: O.is_state_stable ? O.height : 0,
              width: O.width
            }))
          ))
        );
      },
      isPageReload: r,
      onMounted(q) {
        p(B), o(L), a(M), n(P), d.setState(q);
      },
      onUnmounted() {
        F(B), c(L), m(M), i(P);
      },
      restoreState: t.get
    }), G = (q) => C(() => this.safeAreaInsets()[q]), ae = (q) => C(() => this.contentSafeAreaInsets()[q]);
    this.state = d.state, this.height = d.getter("height"), this.stableHeight = d.getter("stableHeight"), this.width = d.getter("width"), this.isExpanded = d.getter("isExpanded"), this.safeAreaInsets = d.getter("safeAreaInsets"), this.safeAreaInsetTop = G("top"), this.safeAreaInsetBottom = G("bottom"), this.safeAreaInsetLeft = G("left"), this.safeAreaInsetRight = G("right"), this.contentSafeAreaInsets = d.getter("contentSafeAreaInsets"), this.contentSafeAreaInsetTop = ae("top"), this.contentSafeAreaInsetBottom = ae("bottom"), this.contentSafeAreaInsetLeft = ae("left"), this.contentSafeAreaInsetRight = ae("right");
    const st = E({ isTma: k, returns: "task" }), Te = E({
      isTma: k,
      returns: "either"
    }), rt = E({
      isTma: k,
      requires: "web_app_request_fullscreen",
      version: I,
      returns: "task"
    }), Le = (q) => rt((Z) => u.pipe(
      _(
        q ? "web_app_request_fullscreen" : "web_app_exit_fullscreen",
        ["fullscreen_changed", "fullscreen_failed"],
        Z
      ),
      l.chain((x) => x.event === "fullscreen_failed" && x.payload.error !== "ALREADY_FULLSCREEN" ? l.left(new qt(x.payload.error)) : (d.setState({
        isFullscreen: "is_fullscreen" in x.payload ? x.payload.is_fullscreen : !0
      }), l.right(void 0)))
    ));
    this.isMounted = R.isMounted, this.mountFp = st(R.mount), this.mount = /* @__PURE__ */ h(this.mountFp), this.isFullscreen = d.getter("isFullscreen"), this.requestFullscreenFp = Le(!0), this.requestFullscreen = /* @__PURE__ */ h(this.requestFullscreenFp), this.exitFullscreenFp = Le(!1), this.exitFullscreen = /* @__PURE__ */ h(this.exitFullscreenFp);
    const pe = j(!1);
    this.isCssVarsBound = C(pe), this.bindCssVarsFp = Te(
      (q) => {
        if (pe())
          return b.left(new ye());
        q || (q = (x) => `--tg-viewport-${Ge(x)}`);
        const Z = [
          ["height", this.height],
          ["stableHeight", this.stableHeight],
          ["width", this.width],
          ["safeAreaInsetTop", this.safeAreaInsetTop],
          ["safeAreaInsetBottom", this.safeAreaInsetBottom],
          ["safeAreaInsetLeft", this.safeAreaInsetLeft],
          ["safeAreaInsetRight", this.safeAreaInsetRight],
          ["contentSafeAreaInsetTop", this.contentSafeAreaInsetTop],
          ["contentSafeAreaInsetBottom", this.contentSafeAreaInsetBottom],
          ["contentSafeAreaInsetLeft", this.contentSafeAreaInsetLeft],
          ["contentSafeAreaInsetRight", this.contentSafeAreaInsetRight]
        ].reduce((x, [H, N]) => {
          const O = q(H);
          if (O) {
            const $e = () => {
              Me(O, `${N()}px`);
            };
            x.push({ update: $e, removeListener: N.sub($e), cssVar: O });
          }
          return x;
        }, []);
        return Z.forEach((x) => {
          x.update();
        }), pe.set(!0), b.right(() => {
          Z.forEach((x) => {
            x.removeListener(), Ve(x.cssVar);
          }), pe.set(!1);
        });
      }
    ), this.bindCssVars = /* @__PURE__ */ h(this.bindCssVarsFp), this.expandFp = Te(() => f("web_app_expand")), this.expand = /* @__PURE__ */ h(this.expandFp);
  }
  //#endregion
}
function Sr() {
  const s = (p) => ({
    on: (o) => {
      z(p, o);
    },
    off: (o) => {
      ne(p, o);
    }
  }), t = s("viewport_changed"), r = s("fullscreen_changed"), n = s("safe_area_changed"), a = s("content_safe_area_changed");
  return new wr({
    ...u.pipe(
      v(),
      Y("viewport"),
      y,
      T,
      $
    ),
    isFullscreen() {
      return u.pipe(le(), b.map((p) => !!p.tgWebAppFullscreen));
    },
    isViewportStable() {
      return u.pipe(le(), b.map((p) => ["macos", "tdesktop", "unigram", "webk", "weba", "web"].includes(p.tgWebAppPlatform)));
    },
    offContentSafeAreaInsetsChanged: a.off,
    offFullscreenChanged: r.off,
    offSafeAreaInsetsChanged: n.off,
    offViewportChanged: t.off,
    onContentSafeAreaInsetsChanged: a.on,
    onFullscreenChanged: r.on,
    onSafeAreaInsetsChanged: n.on,
    onViewportChanged: t.on
  });
}
const Fn = /* @__PURE__ */ Sr();
function vr(s = {}) {
  const {
    version: t,
    isInlineMode: r,
    themeParams: n
  } = s;
  if (t && typeof r == "boolean" && n)
    we.set(t), ve.set(r), Se.set(n);
  else {
    const i = u.pipe(le(), b.matchW(
      (c) => c,
      (c) => {
        we.set(t || c.tgWebAppVersion), ve.set(typeof r == "boolean" ? r : !!c.tgWebAppBotInline), Se.set(n || c.tgWebAppThemeParams);
      }
    ));
    if (i)
      return b.left(i);
  }
  s.postEvent && ze.set(s.postEvent);
  const [a, p] = ke(
    z("reload_iframe", () => {
      Re().log("Received a request to reload the page"), Mt("iframe_will_reload"), window.location.reload();
    })
  ), { acceptCustomStyles: o = !0 } = s;
  if (o) {
    const i = document.createElement("style");
    i.id = "telegram-custom-styles", document.head.appendChild(i), a(
      z("set_custom_style", (c) => {
        i.innerHTML = c;
      }),
      () => {
        document.head.removeChild(i);
      }
    );
  }
  return u.pipe(
    re("iframe_ready", { reload_supported: !0 }),
    b.map(() => (Re().log("The package was initialized"), p))
  );
}
const Cn = te(vr);
export {
  Qe as AccessDeniedError,
  xt as BackButton,
  $t as Biometry,
  ye as CSSVarsBoundError,
  vn as CancelledError,
  jt as ClosingBehavior,
  Dt as CloudStorage,
  Ee as ConcurrentCallError,
  Wt as DeviceStorage,
  vt as DeviceStorageMethodError,
  qt as FullscreenFailedError,
  yt as FunctionUnavailableError,
  Zt as HapticFeedback,
  os as InitData,
  U as InvalidArgumentsError,
  Mr as InvalidEnvError,
  kn as InvalidLaunchParamsError,
  ps as Invoice,
  In as InvokeCustomMethodFailedError,
  yn as LaunchParamsRetrieveError,
  Fs as LocationManager,
  ws as MainButton,
  En as MethodParameterUnsupportedError,
  qn as MethodUnsupportedError,
  Is as MiniApp,
  It as NotAvailableError,
  As as Popup,
  Ds as QrScanner,
  Us as SecondaryButton,
  Hs as SecureStorage,
  kt as SecureStorageMethodError,
  Et as SetEmojiStatusError,
  Qs as SettingsButton,
  At as ShareMessageError,
  zs as SwipeBehavior,
  vs as ThemeParams,
  An as TimeoutError,
  Bn as UnknownEnvError,
  Bt as UnknownThemeParamsKeyError,
  ue as ValidationError,
  wr as Viewport,
  Ur as addToHomeScreen,
  ss as addToHomeScreenFp,
  xn as applyPolyfills,
  Vr as backButton,
  $r as biometry,
  Wr as checkHomeScreenStatus,
  is as checkHomeScreenStatusFp,
  Rr as closingBehavior,
  Pr as cloudStorage,
  un as copyTextToClipboard,
  Mn as createLogger,
  Vn as createPostEvent,
  Be as createRequestId,
  Tn as createStartParam,
  Ln as createStartParamFp,
  $n as debug,
  Rn as decodeBase64Url,
  Pn as decodeBase64UrlFp,
  jn as decodeStartParam,
  On as decodeStartParamFp,
  Dn as deepSnakeToCamelObjKeys,
  jr as deviceStorage,
  cn as downloadFile,
  Zs as downloadFileFp,
  Gn as emitEvent,
  Un as encodeBase64Url,
  ln as getCurrentTime,
  tr as getCurrentTimeFp,
  Wn as getReleaseVersion,
  Gr as hapticFeedback,
  hn as hideKeyboard,
  nr as hideKeyboardFp,
  Cn as init,
  Hr as initData,
  vr as initFp,
  Kr as invoice,
  Gt as invokeCustomMethod,
  ci as isAnyRGB,
  Ze as isColorDark,
  Ss as isColorDarkFp,
  li as isRGB,
  hi as isRGBA,
  di as isRGBAShort,
  _i as isRGBShort,
  Hn as isSafeToCreateStartParam,
  Kn as isTMA,
  Qn as isTMAFp,
  Jr as locationManager,
  zn as logger,
  Yr as mainButton,
  Es as miniApp,
  Nn as mockTelegramEnv,
  Jn as off,
  Yn as offAll,
  Zn as on,
  Qr as openLink,
  hs as openLinkFp,
  zr as openTelegramLink,
  Ye as openTelegramLinkFp,
  mi as parseInitDataQuery,
  fi as parseInitDataQueryFp,
  bi as parseLaunchParamsQuery,
  gi as parseLaunchParamsQueryFp,
  Zr as popup,
  Mt as postEvent,
  re as postEventFp,
  rn as qrScanner,
  dn as readTextFromClipboard,
  ar as readTextFromClipboardFp,
  Tr as request,
  Lr as request2,
  Ne as request2Fp,
  tn as requestContact,
  en as requestContactComplete,
  tt as requestContactCompleteFp,
  Rs as requestContactFp,
  Or as requestEmojiStatusAccess,
  zt as requestEmojiStatusAccessFp,
  Rt as requestFp,
  Xr as requestPhoneAccess,
  et as requestPhoneAccessFp,
  sn as requestWriteAccess,
  Os as requestWriteAccessFp,
  _n as retrieveAndroidDeviceData,
  pr as retrieveAndroidDeviceDataFrom,
  Xn as retrieveLaunchParams,
  ei as retrieveLaunchParamsFp,
  ti as retrieveRawInitData,
  si as retrieveRawInitDataFp,
  ri as retrieveRawLaunchParams,
  ni as retrieveRawLaunchParamsFp,
  nn as secondaryButton,
  on as secureStorage,
  mn as sendData,
  lr as sendDataFp,
  Fi as serializeInitDataQuery,
  Ci as serializeLaunchParamsQuery,
  ii as setDebug,
  Dr as setEmojiStatus,
  Yt as setEmojiStatusFp,
  oi as setTargetOrigin,
  an as settingsButton,
  fn as shareMessage,
  _r as shareMessageFp,
  bn as shareStory,
  br as shareStoryFp,
  Nr as shareURL,
  bs as shareURLFp,
  ai as supports,
  pn as swipeBehavior,
  gn as switchInlineQuery,
  Cr as switchInlineQueryFp,
  pi as targetOrigin,
  he as themeParams,
  wi as toRGB,
  Si as toRGBFp,
  vi as toRGBFull,
  ki as toRGBFullFp,
  Fn as viewport
};
