// @tma.js/bridge/dist/index.js
import * as T from "valibot";
import { looseObject as d, nullish as v, boolean as z, number as V, string as E, optional as B, unknown as we, parse as L, is as O, function as de, any as G } from "valibot";
import { throwifyFpFn as A, getStorageValue as be, setStorageValue as Z, createLogger as ge, throwifyAnyEither as C, createCbCollector as ee, BetterTaskEither as te } from "@tma.js/toolkit";
import { createLogger as vt, deepSnakeToCamelObjKeys as Et } from "@tma.js/toolkit";
import { BetterPromise as re, TimeoutError as he } from "better-promises";
import { CancelledError as xt, TimeoutError as kt } from "better-promises";
import { function as l, either as c, option as R, taskEither as f, json as ye } from "fp-ts";
import { errorClassWithData as ve, errorClass as x } from "error-kid";
import { parseLaunchParamsQueryFp as ne, themeParams as Ee, pipeJsonToSchema as oe, miniAppsMessage as ae, serializeLaunchParamsQuery as Pe, parseLaunchParamsQuery as xe } from "@tma.js/transformers";
import { signal as g, computed as se } from "@tma.js/signals";
import ke from "mitt";
function ie(e) {
  return typeof e == "object" && !!e && !Array.isArray(e) && T.is(
    T.looseObject({ TelegramWebviewProxy: T.looseObject({ postEvent: T.function() }) }),
    { TelegramWebviewProxy: e.TelegramWebviewProxy }
  );
}
function pe() {
  try {
    return window.self !== window.top;
  } catch {
    return !0;
  }
}
class Te extends (/* @__PURE__ */ x({
  name: "MethodUnsupportedError",
  super: (t, r) => [
    `Method "${t}" is unsupported in Mini Apps version ${r}`
  ]
})) {
}
class Se extends (/* @__PURE__ */ x({
  name: "MethodParameterUnsupportedError",
  super: (t, r, n) => [
    `Parameter "${r}" of "${t}" method is unsupported in Mini Apps version ${n}`
  ]
})) {
}
class Ae extends (/* @__PURE__ */ ve({
  name: "LaunchParamsRetrieveError",
  data: (t) => ({ errors: t }),
  super: (t) => [
    [
      "Unable to retrieve launch parameters from any known source. Perhaps, you have opened your app outside Telegram?",
      "📖 Refer to docs for more information:",
      "https://docs.telegram-mini-apps.com/packages/tma-js-bridge/environment",
      "",
      "Collected errors:",
      ...t.map(({ source: r, error: n }) => `Source: ${r} / ${n instanceof Error ? n.message : String(n)}`)
    ].join(`
`)
  ]
})) {
}
class Ce extends (/* @__PURE__ */ x({
  name: "InvalidLaunchParamsError",
  super: (t, r) => [
    `Invalid value for launch params: ${t}`,
    { cause: r }
  ]
})) {
}
class ce extends (/* @__PURE__ */ x({ name: "UnknownEnvError" })) {
}
class Me extends (/* @__PURE__ */ x({
  name: "InvokeCustomMethodError",
  super: (t) => [`Server returned error: ${t}`]
})) {
}
const D = "launchParams";
function J(e) {
  return e.replace(/^[^?#]*[?#]/, "").replace(/[?#]/g, "&");
}
function We() {
  return l.pipe(
    M(),
    c.chainW(ne)
  );
}
const rt = A(We);
function je() {
  return l.pipe(
    M(),
    c.map((e) => {
      const t = new URLSearchParams(e).get("tgWebAppData");
      return t ? R.some(t) : R.none;
    })
  );
}
function nt() {
  return l.pipe(
    je(),
    c.fold((e) => {
      throw e;
    }, (e) => e),
    R.match(() => {
    }, (e) => e)
  );
}
function M() {
  const e = [];
  for (const [t, r] of [
    // Try to retrieve launch parameters from the current location. This method
    // can return nothing in case, location was changed, and then the page was
    // reloaded.
    [() => J(window.location.href), "window.location.href"],
    // Then, try using the lower level API - window.performance.
    [() => {
      const n = performance.getEntriesByType("navigation")[0];
      return n && J(n.name);
    }, "performance navigation entries"],
    // Finally, try using the session storage.
    [() => be(D), "local storage"]
  ]) {
    const n = t();
    if (!n) {
      e.push({ source: r, error: new Error("Source is empty") });
      continue;
    }
    const a = l.pipe(
      ne(n),
      c.foldW((o) => o, () => !0)
    );
    if (typeof a != "boolean") {
      e.push({ source: r, error: a });
      continue;
    }
    return Z(D, n), c.right(n);
  }
  return c.left(new Ae(e));
}
const ot = A(M);
function qe(e, t) {
  const r = /* @__PURE__ */ new Map(), n = ke(), a = (o, s, p = !1) => {
    const _ = r.get(o) || /* @__PURE__ */ new Map();
    r.set(o, _);
    const i = _.get(s) || [];
    _.set(s, i);
    const u = i.findIndex((m) => m[1] === p);
    if (u >= 0 && (n.off(o, i[u][0]), i.splice(u, 1), !i.length && (_.delete(s), !_.size))) {
      const m = r.size;
      r.delete(o), m && !r.size && t();
    }
  };
  return {
    on(o, s, p) {
      !r.size && e();
      const _ = () => {
        a(o, s, p);
      }, i = (...w) => {
        p && _(), o === "*" ? s({ name: w[0], payload: w[1] }) : s(...w);
      };
      n.on(o, i);
      const u = r.get(o) || /* @__PURE__ */ new Map();
      r.set(o, u);
      const m = u.get(s) || [];
      return u.set(s, m), m.push([i, p || !1]), _;
    },
    off: a,
    // eslint-disable-next-line @typescript-eslint/unbound-method
    emit: n.emit,
    clear() {
      const o = r.size;
      n.all.clear(), r.clear(), o && t();
    }
  };
}
function j(e, t) {
  window.dispatchEvent(new MessageEvent("message", {
    data: JSON.stringify({ eventType: e, eventData: t }),
    // We specify this kind of source here in order to allow the package's "on" function to
    // capture it. The reason is this function always checks the event source and relies on
    // it to be the parent window.
    source: window.parent
  }));
}
const S = g(!1), U = g("https://web.telegram.org"), Fe = (e) => {
  b().log("Event received:", e);
}, _e = se(S);
function at(e) {
  e !== S() && (S.set(e), (e ? I : Oe)("*", Fe));
}
const ue = se(U);
function st(e) {
  U.set(e), b().log("New target origin set", e);
}
const P = g((...e) => {
  window.parent.postMessage(...e);
}), b = g(ge("Bridge", {
  bgColor: "#9147ff",
  textColor: "white",
  shouldLog: _e
}));
function it() {
  Ue(), [P, U, ue, S, _e, b].forEach((e) => {
    e.unsubAll(), "reset" in e && e.reset();
  });
}
function q(e, t, r) {
  const n = [r], a = e[t];
  typeof a == "function" && n.push(a);
  const o = (...p) => {
    n.forEach((_) => _(...p));
  }, s = Object.assign((...p) => {
    o(...p);
  }, {
    // Unwraps the composer.
    unwrap() {
      const { length: p } = n;
      if (p === 1) {
        delete e[t];
        return;
      }
      if (p === 2) {
        Q(e, t, n[1]);
        return;
      }
      n.unshift(1), Q(e, t, o);
    }
  });
  me(
    e,
    t,
    () => s,
    (p) => n.push(p)
  );
}
function N(e, t) {
  const r = e[t];
  me(e, t, () => r, (n) => {
    Object.entries(n).forEach(([a, o]) => {
      r[a] = o;
    });
  });
}
function me(e, t, r, n) {
  Object.defineProperty(e, t, {
    enumerable: !0,
    configurable: !0,
    get: r,
    set: n
  });
}
function Q(e, t, r) {
  Object.defineProperty(e, t, {
    enumerable: !0,
    configurable: !0,
    writable: !0,
    value: r
  });
}
const Le = {
  clipboard_text_received: d({
    req_id: E(),
    data: v(E())
  }),
  custom_method_invoked: d({
    req_id: E(),
    result: B(we()),
    error: B(E())
  }),
  popup_closed: v(
    d({ button_id: v(E(), () => {
    }) }),
    {}
  ),
  viewport_changed: v(
    d({
      height: V(),
      width: v(V(), () => window.innerWidth),
      is_state_stable: z(),
      is_expanded: z()
    }),
    // TODO: At the moment, macOS has a bug with the invalid event payload - it is always equal to
    //  null. Leaving this default value until the bug is fixed.
    () => ({
      height: window.innerHeight,
      is_state_stable: !0,
      is_expanded: !0
    })
  ),
  theme_changed: d({
    theme_params: Ee()
  })
};
function H(e) {
  if (e.source !== window.parent)
    return;
  let t;
  try {
    t = L(oe(ae()), e.data);
  } catch {
    return;
  }
  const { eventType: r, eventData: n } = t, a = Le[r];
  let o;
  try {
    o = a ? L(a, n) : n;
  } catch (s) {
    return b().forceError(
      [
        `An error occurred processing the "${r}" event from the Telegram application.`,
        "Please, file an issue here:",
        "https://github.com/Telegram-Mini-Apps/tma.js/issues/new/choose"
      ].join(`
`),
      t,
      s
    );
  }
  Re(r, o);
}
const {
  on: I,
  off: Oe,
  emit: Re,
  clear: Ue
} = qe(
  () => {
    const e = window;
    !e.TelegramGameProxy && (e.TelegramGameProxy = {}), q(e.TelegramGameProxy, "receiveEvent", j), N(e, "TelegramGameProxy"), !e.Telegram && (e.Telegram = {}), !e.Telegram.WebView && (e.Telegram.WebView = {}), q(e.Telegram.WebView, "receiveEvent", j), N(e.Telegram, "WebView"), q(e, "TelegramGameProxy_receiveEvent", j), window.addEventListener("message", H);
  },
  () => {
    [
      ["TelegramGameProxy_receiveEvent"],
      ["TelegramGameProxy", "receiveEvent"],
      ["Telegram", "WebView", "receiveEvent"]
    ].forEach((e) => {
      const t = window;
      let r = [void 0, t];
      for (const o of e)
        if (r = [r[1], r[1][o]], !r[1])
          return;
      const [n, a] = r;
      "unwrap" in a && (a.unwrap(), n && n !== t && !Object.keys(n).length && delete t[e[0]]);
    }), window.removeEventListener("message", H);
  }
), Ie = (...e) => P()(...e);
function $e(e, t) {
  l.pipe(
    k(
      // @ts-expect-error It's ok, TS can't determine a specific override.
      e,
      t
    ),
    c.mapLeft((r) => {
      throw r;
    })
  );
}
function k(e, t) {
  b().log("Posting event:", t ? { eventType: e, eventData: t } : { eventType: e });
  const r = window, n = JSON.stringify({ eventType: e, eventData: t });
  return pe() ? (Ie(n, ue()), c.right(void 0)) : ie(r) ? (r.TelegramWebviewProxy.postEvent(e, JSON.stringify(t)), c.right(void 0)) : O(d({ external: d({ notify: de() }) }), r) ? (r.external.notify(n), c.right(void 0)) : c.left(new ce());
}
function $(e, t, r = {}) {
  const {
    // If no capture function was passed, we capture the first compatible event.
    capture: n = () => !0,
    postEvent: a = k
  } = r, o = g(), [s, p] = ee();
  (Array.isArray(t) ? t : [t]).forEach((i) => {
    s(
      I(i, (u) => {
        const m = Array.isArray(t);
        n(m ? { event: i, payload: u } : u) && o.set([
          m ? { event: i, payload: u } : u
        ]);
      })
    );
  });
  const _ = (i) => (p(), i);
  return l.pipe(
    async () => a(e, r.params),
    f.chainW(() => te((i, u, m) => {
      const w = o();
      if (w)
        return i(w[0]);
      const h = (y) => {
        y && i(y[0]);
      }, W = () => {
        o.unsub(h);
      };
      o.sub(h), m.on("finalized", W);
    }, r)),
    f.mapBoth(_, _)
  );
}
function pt(e, t, r) {
  const { postEvent: n } = r || {};
  return C(
    // @ts-expect-error TypeScript will not be able to handle our overrides here.
    $(e, t, {
      ...r,
      postEvent: n ? (...a) => {
        try {
          return n(...a), c.right(void 0);
        } catch (o) {
          return c.left(o);
        }
      } : k
    })
  );
}
function ct(e, t) {
  const r = ze(
    // @ts-expect-error TS doesn't get what override we are going to use.
    e,
    t
  );
  return typeof r == "function" ? re.fn(() => C(r)) : r;
}
function ze(e, t) {
  const r = ie(window);
  if (!e)
    return r || l.pipe(M(), c.match(() => !1, () => !0));
  if (r)
    return f.right(!0);
  const { timeout: n = 100 } = t || {};
  return l.pipe(
    $("web_app_request_theme", "theme_changed", { ...t, timeout: n }),
    f.match(
      (a) => he.is(a) || ce.is(a) ? c.right(!1) : c.left(a),
      () => c.right(!0)
    )
  );
}
function _t({ launchParams: e, onEvent: t, resetPostMessage: r } = {}) {
  if (e) {
    const o = typeof e == "string" || e instanceof URLSearchParams ? e.toString() : (
      // Here we have to trick serializeLaunchParamsQuery into thinking, it serializes a valid
      // value. We are doing it because we are working with tgWebAppData presented as a
      // string, not an object as serializeLaunchParamsQuery requires.
      Pe({ ...e, tgWebAppData: void 0 }) + (e.tgWebAppData ? `&tgWebAppData=${encodeURIComponent(e.tgWebAppData.toString())}` : "")
    );
    try {
      xe(o);
    } catch (s) {
      throw new Ce(o, s);
    }
    Z("launchParams", o);
  }
  if (pe()) {
    if (!t)
      return;
    r && P.reset();
    const o = P();
    P.set((...s) => {
      const [p] = s, _ = () => {
        o(...s);
      };
      try {
        const i = L(oe(ae()), p);
        t({ name: i.eventType, params: i.eventData }, _);
      } catch {
        _();
      }
    });
    return;
  }
  const n = window.TelegramWebviewProxy || {}, a = n.postEvent || (() => {
  });
  window.TelegramWebviewProxy = {
    ...n,
    postEvent(o, s) {
      const p = () => {
        a(o, s);
      };
      t ? t({
        name: o,
        params: s ? JSON.parse(s) : void 0
      }, p) : p();
    }
  }, b().log("Environment was mocked by the mockTelegramEnv function");
}
function Ve(e) {
  return ({ req_id: t }) => t === e;
}
const K = {
  "6.0": [
    "iframe_ready",
    "iframe_will_reload",
    "web_app_close",
    "web_app_data_send",
    "web_app_expand",
    "web_app_open_link",
    "web_app_ready",
    "web_app_request_theme",
    "web_app_request_viewport",
    "web_app_setup_main_button",
    "web_app_setup_closing_behavior"
  ],
  6.1: [
    "web_app_open_tg_link",
    "web_app_open_invoice",
    "web_app_setup_back_button",
    "web_app_set_background_color",
    "web_app_set_header_color",
    "web_app_trigger_haptic_feedback"
  ],
  6.2: ["web_app_open_popup"],
  6.4: [
    "web_app_close_scan_qr_popup",
    "web_app_open_scan_qr_popup",
    "web_app_read_text_from_clipboard",
    { method: "web_app_open_link", param: "try_instant_view" }
  ],
  6.7: ["web_app_switch_inline_query"],
  6.9: [
    "web_app_invoke_custom_method",
    "web_app_request_write_access",
    "web_app_request_phone",
    { method: "web_app_set_header_color", param: "color" }
  ],
  "6.10": ["web_app_setup_settings_button"],
  7.2: [
    "web_app_biometry_get_info",
    "web_app_biometry_open_settings",
    "web_app_biometry_request_access",
    "web_app_biometry_request_auth",
    "web_app_biometry_update_token"
  ],
  7.6: [
    { method: "web_app_open_link", param: "try_browser" },
    { method: "web_app_close", param: "return_back" }
  ],
  7.7: ["web_app_setup_swipe_behavior"],
  7.8: ["web_app_share_to_story"],
  "7.10": [
    "web_app_setup_secondary_button",
    "web_app_set_bottom_bar_color",
    { method: "web_app_setup_main_button", param: "has_shine_effect" }
  ],
  "8.0": [
    "web_app_request_safe_area",
    "web_app_request_content_safe_area",
    "web_app_request_fullscreen",
    "web_app_exit_fullscreen",
    "web_app_set_emoji_status",
    "web_app_add_to_home_screen",
    "web_app_check_home_screen",
    "web_app_request_emoji_status_access",
    "web_app_check_location",
    "web_app_open_location_settings",
    "web_app_request_file_download",
    "web_app_request_location",
    "web_app_send_prepared_message",
    "web_app_start_accelerometer",
    "web_app_start_device_orientation",
    "web_app_start_gyroscope",
    "web_app_stop_accelerometer",
    "web_app_stop_device_orientation",
    "web_app_stop_gyroscope",
    "web_app_toggle_orientation_lock"
  ],
  "9.0": [
    "web_app_device_storage_clear",
    "web_app_device_storage_get_key",
    "web_app_device_storage_save_key",
    "web_app_secure_storage_clear",
    "web_app_secure_storage_get_key",
    "web_app_secure_storage_restore_key",
    "web_app_secure_storage_save_key"
  ],
  9.1: ["web_app_hide_keyboard"],
  9.5: [
    { method: "web_app_setup_main_button", param: "icon_custom_emoji_id" },
    { method: "web_app_setup_secondary_button", param: "icon_custom_emoji_id" }
  ]
};
function Y(e, t) {
  return Object.keys(K).find((n) => K[n].some((a) => t ? typeof a == "object" && a.method === e && a.param === t : a === e)) || null;
}
function X(e) {
  return e.split(".").map(Number);
}
function Be(e, t) {
  const r = X(e), n = X(t), a = Math.max(r.length, n.length);
  for (let o = 0; o < a; o += 1) {
    const s = r[o] || 0, p = n[o] || 0;
    if (s !== p)
      return s > p ? 1 : -1;
  }
  return 0;
}
function F(e, t, r) {
  const n = r ? Y(
    e,
    t
  ) : Y(e);
  return n ? Be(n, r || t) <= 0 : !1;
}
function ut(e, t = "strict") {
  const r = typeof t == "function" ? t : (n) => {
    const { method: a, version: o } = n, s = "param" in n ? new Se(a, n.param, o) : new Te(a, o);
    if (t === "strict")
      throw s;
    return b().forceWarn(s.message);
  };
  return (n, a) => F(n, e) ? n === "web_app_set_header_color" && O(d({ color: G() }), a) && !F(n, "color", e) ? r({ version: e, method: n, param: "color" }) : (n === "web_app_setup_main_button" || n === "web_app_setup_secondary_button") && O(d({ icon_custom_emoji_id: G() }), a) && !F(n, "icon_custom_emoji_id", e) ? r({ version: e, method: n, param: "icon_custom_emoji_id" }) : $e(n, a) : r({ version: e, method: n });
}
function Ge(e, t, r, n) {
  return l.pipe(
    $("web_app_invoke_custom_method", "custom_method_invoked", {
      ...n || {},
      params: { method: e, params: t, req_id: r },
      capture: Ve(r)
    }),
    f.chain(({ result: a, error: o }) => o ? f.left(new Me(o)) : f.right(a))
  );
}
function mt(e, t, r, n) {
  return re.fn(() => l.pipe(
    // @ts-expect-error TypeScript is unable to determine required override.
    Ge(e, t, r, n),
    f.match(
      (a) => {
        throw a;
      },
      (a) => a
    )
  )());
}
function De(e, t, r = {}) {
  const {
    // If no capture function was passed, we capture the first compatible event.
    capture: n = () => !0,
    postEvent: a = k
  } = r, o = g(), [s, p] = ee();
  (Array.isArray(t) ? t : [t]).forEach((i) => {
    s(
      I(i, (u) => {
        (Array.isArray(t) ? n({ event: i, payload: u }) : n(u)) && o.set([u]);
      })
    );
  });
  const _ = (i) => (p(), i);
  return l.pipe(
    async () => a(e, r.params),
    f.chainW(() => te((i, u, m) => {
      const w = o();
      if (w)
        return i(w[0]);
      const h = (y) => {
        y && i(y[0]);
      }, W = () => {
        o.unsub(h);
      };
      o.sub(h), m.on("finalized", W);
    }, r)),
    f.mapBoth(_, _)
  );
}
function lt(e, t, r) {
  const { postEvent: n } = r || {};
  return C(
    // @ts-expect-error TypeScript will not be able to handle our overrides here.
    De(e, t, {
      ...r,
      postEvent: n ? (...a) => {
        try {
          return n(...a), c.right(void 0);
        } catch (o) {
          return c.left(o);
        }
      } : k
    })
  );
}
function ft() {
  Object.hasOwn || (Object.hasOwn = function(e, t) {
    return Object.prototype.hasOwnProperty.call(e, t);
  });
}
function le(e) {
  return c.tryCatch(() => decodeURIComponent(
    atob(
      e.replace(/-/g, "+").replace(/_/g, "/")
    ).split("").map((t) => "%" + ("00" + t.charCodeAt(0).toString(16)).slice(-2)).join("")
  ), (t) => t);
}
const wt = A(le);
function fe(e) {
  return btoa(
    encodeURIComponent(e).replace(/%([0-9A-F]{2})/g, (t, r) => String.fromCharCode(parseInt(`0x${r}`)))
  ).replace(/\+/g, "-").replace(/\//g, "_");
}
function Je(e) {
  const t = fe(typeof e == "string" ? e : JSON.stringify(e));
  return t.length > 512 ? c.left(new Error("Value is too long for start parameter")) : c.right(t);
}
const dt = A(Je);
function bt(e, t) {
  return C(
    Ne(
      e,
      // @ts-expect-error TypeScript is unable to detect a correct override.
      typeof t == "function" ? (r) => c.tryCatch(() => t(r), (n) => n) : t
    )
  );
}
function Ne(e, t) {
  return l.pipe(
    le(e),
    c.chain((r) => t ? typeof t == "function" ? t(r) : ye.parse(r) : c.right(r))
  );
}
function gt(e) {
  return fe(e).length <= 512;
}
export {
  xt as CancelledError,
  Ce as InvalidLaunchParamsError,
  Me as InvokeCustomMethodFailedError,
  Ae as LaunchParamsRetrieveError,
  Se as MethodParameterUnsupportedError,
  Te as MethodUnsupportedError,
  kt as TimeoutError,
  ce as UnknownEnvError,
  ft as applyPolyfills,
  Ve as captureSameReq,
  Be as compareVersions,
  vt as createLogger,
  ut as createPostEvent,
  dt as createStartParam,
  Je as createStartParamFp,
  _e as debug,
  wt as decodeBase64Url,
  le as decodeBase64UrlFp,
  bt as decodeStartParam,
  Ne as decodeStartParamFp,
  Et as deepSnakeToCamelObjKeys,
  j as emitEvent,
  fe as encodeBase64Url,
  Y as getReleaseVersion,
  ie as hasWebviewProxy,
  mt as invokeCustomMethod,
  Ge as invokeCustomMethodFp,
  pe as isIframe,
  gt as isSafeToCreateStartParam,
  ct as isTMA,
  ze as isTMAFp,
  b as logger,
  _t as mockTelegramEnv,
  Oe as off,
  Ue as offAll,
  I as on,
  $e as postEvent,
  k as postEventFp,
  Ie as postMessage,
  P as postMessageImpl,
  lt as request,
  pt as request2,
  $ as request2Fp,
  De as requestFp,
  it as resetGlobals,
  rt as retrieveLaunchParams,
  We as retrieveLaunchParamsFp,
  nt as retrieveRawInitData,
  je as retrieveRawInitDataFp,
  ot as retrieveRawLaunchParams,
  M as retrieveRawLaunchParamsFp,
  at as setDebug,
  st as setTargetOrigin,
  F as supports,
  ue as targetOrigin
};
